<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Compile model — compile_model • deepG</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Compile model — compile_model"><meta name="description" content="Compile model"><meta property="og:description" content="Compile model"><meta property="og:image" content="https://genomenet.github.io/deepG/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">deepG</a>

    <small class="nav-text text-default me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.3.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html"><span class="fa fa-file-alt"></span> Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-notebooks" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Notebooks</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-notebooks"><li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/175jIdXcDcgPUvaBo2rH2Lupbpjnp5O7G?usp=sharing">deepG tutorial</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/1Eolc0koMNM1zkuO4XyVM58ImeF1BpRiH?usp=sharing">Read-length level: Human contamination</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/1yiXSwFafXpMLHaov9iBTQLIDZ6bK1zYX?usp=sharing">Locus level: CRISPR detection</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/1G7bOFEX87cZNrM2tdRtTdkrZn5fM__g0?usp=sharing">Gene level: 16S rRNA detection</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/1BCggL-tfQF136YeJ8cKKi-zoBEDMgkNh?usp=sharing">Genome level: Bacterial morphology (Sporulation)</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/10xpRzGd3JeBAbqQYSCxzQUMctt01sx9D?usp=sharing">Full metagenome level: Colorectal cancer prediction</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/1kyYK7IU7GSfdpDzO_a8U3_qD4i3zTu6w?usp=sharing">BERT with deepG</a></li>
  </ul></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-tutorials" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Tutorials</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-tutorials"><li><a class="dropdown-item" href="../articles/getting_started.html">Getting Started</a></li>
    <li><a class="dropdown-item" href="../articles/training_types.html">Training types</a></li>
    <li><a class="dropdown-item" href="../articles/data_generator.html">Data generator</a></li>
    <li><a class="dropdown-item" href="../articles/using_tb.html">Using tensorboard</a></li>
    <li><a class="dropdown-item" href="../articles/integrated_gradient.html">Integrated Gradient</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/GenomeNet/deepG/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Compile model</h1>
      <small class="dont-index">Source: <a href="https://github.com/GenomeNet/deepG/blob/master/R/create_model_utils.R" class="external-link"><code>R/create_model_utils.R</code></a></small>
      <div class="d-none name"><code>compile_model.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Compile model</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">compile_model</span><span class="op">(</span></span>
<span>  <span class="va">model</span>,</span>
<span>  <span class="va">solver</span>,</span>
<span>  <span class="va">learning_rate</span>,</span>
<span>  <span class="va">loss_fn</span>,</span>
<span>  label_smoothing <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  num_output_layers <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  label_noise_matrix <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bal_acc <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  f1_metric <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  auc_metric <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  layer_dense <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-model">model<a class="anchor" aria-label="anchor" href="#arg-model"></a></dt>
<dd><p>A keras model.</p></dd>


<dt id="arg-solver">solver<a class="anchor" aria-label="anchor" href="#arg-solver"></a></dt>
<dd><p>Optimization method, options are <code>"adam", "adagrad", "rmsprop"</code> or <code>"sgd"</code>.</p></dd>


<dt id="arg-learning-rate">learning_rate<a class="anchor" aria-label="anchor" href="#arg-learning-rate"></a></dt>
<dd><p>Learning rate for optimizer.</p></dd>


<dt id="arg-loss-fn">loss_fn<a class="anchor" aria-label="anchor" href="#arg-loss-fn"></a></dt>
<dd><p>Either <code>"categorical_crossentropy"</code> or <code>"binary_crossentropy"</code>. If <code>label_noise_matrix</code> given, will use custom <code>"noisy_loss"</code>.</p></dd>


<dt id="arg-label-smoothing">label_smoothing<a class="anchor" aria-label="anchor" href="#arg-label-smoothing"></a></dt>
<dd><p>Float in [0, 1]. If 0, no smoothing is applied. If &gt; 0, loss between the predicted
labels and a smoothed version of the true labels, where the smoothing squeezes the labels towards 0.5.
The closer the argument is to 1 the more the labels get smoothed.</p></dd>


<dt id="arg-num-output-layers">num_output_layers<a class="anchor" aria-label="anchor" href="#arg-num-output-layers"></a></dt>
<dd><p>Number of output layers.</p></dd>


<dt id="arg-label-noise-matrix">label_noise_matrix<a class="anchor" aria-label="anchor" href="#arg-label-noise-matrix"></a></dt>
<dd><p>Matrix of label noises. Every row stands for one class and columns for percentage of labels in that class.
If first label contains 5 percent wrong labels and second label no noise, then</p>
<p><code>label_noise_matrix &lt;- matrix(c(0.95, 0.05, 0, 1), nrow = 2, byrow = TRUE )</code></p></dd>


<dt id="arg-bal-acc">bal_acc<a class="anchor" aria-label="anchor" href="#arg-bal-acc"></a></dt>
<dd><p>Whether to add balanced accuracy.</p></dd>


<dt id="arg-f-metric">f1_metric<a class="anchor" aria-label="anchor" href="#arg-f-metric"></a></dt>
<dd><p>Whether to add F1 metric.</p></dd>


<dt id="arg-auc-metric">auc_metric<a class="anchor" aria-label="anchor" href="#arg-auc-metric"></a></dt>
<dd><p>Whether to add AUC metric.</p></dd>


<dt id="arg-layer-dense">layer_dense<a class="anchor" aria-label="anchor" href="#arg-layer-dense"></a></dt>
<dd><p>Vector specifying number of neurons per dense layer after last LSTM or CNN layer (if no LSTM used).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A compiled keras model.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># reticulate::py_module_available("tensorflow")</span></span></span>
<span class="r-in"><span><span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="create_model_lstm_cnn.html">create_model_lstm_cnn</a></span><span class="op">(</span>layer_lstm <span class="op">=</span> <span class="fl">8</span>, compile <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">model</span> <span class="op">&lt;-</span> <span class="fu">compile_model</span><span class="op">(</span>model <span class="op">=</span> <span class="va">model</span>,</span></span>
<span class="r-in"><span>                       solver <span class="op">=</span> <span class="st">'adam'</span>,</span></span>
<span class="r-in"><span>                       learning_rate <span class="op">=</span> <span class="fl">0.01</span>,</span></span>
<span class="r-in"><span>                       loss_fn <span class="op">=</span> <span class="st">'categorical_crossentropy'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Philipp Münch, René Mreches, Martin Binder, Hüseyin Anil Gündüz, Xiao-Yin To, Alice McHardy.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

