<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Data generator for fasta/fastq files and label targets — generator_fasta_label_header_csv • deepG</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Data generator for fasta/fastq files and label targets — generator_fasta_label_header_csv"><meta name="description" content="Iterates over folder containing fasta/fastq files and produces encoding of predictor sequences
and target variables. Targets will be read from fasta headers or a separate csv file."><meta property="og:description" content="Iterates over folder containing fasta/fastq files and produces encoding of predictor sequences
and target variables. Targets will be read from fasta headers or a separate csv file."><meta property="og:image" content="https://genomenet.github.io/deepG/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">deepG</a>

    <small class="nav-text text-default me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.3.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html"><span class="fa fa-file-alt"></span> Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-notebooks" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Notebooks</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-notebooks"><li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/175jIdXcDcgPUvaBo2rH2Lupbpjnp5O7G?usp=sharing">deepG tutorial</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/1Eolc0koMNM1zkuO4XyVM58ImeF1BpRiH?usp=sharing">Read-length level: Human contamination</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/1yiXSwFafXpMLHaov9iBTQLIDZ6bK1zYX?usp=sharing">Locus level: CRISPR detection</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/1G7bOFEX87cZNrM2tdRtTdkrZn5fM__g0?usp=sharing">Gene level: 16S rRNA detection</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/1BCggL-tfQF136YeJ8cKKi-zoBEDMgkNh?usp=sharing">Genome level: Bacterial morphology (Sporulation)</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/10xpRzGd3JeBAbqQYSCxzQUMctt01sx9D?usp=sharing">Full metagenome level: Colorectal cancer prediction</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/1kyYK7IU7GSfdpDzO_a8U3_qD4i3zTu6w?usp=sharing">BERT with deepG</a></li>
  </ul></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-tutorials" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Tutorials</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-tutorials"><li><a class="dropdown-item" href="../articles/getting_started.html">Getting Started</a></li>
    <li><a class="dropdown-item" href="../articles/training_types.html">Training types</a></li>
    <li><a class="dropdown-item" href="../articles/data_generator.html">Data generator</a></li>
    <li><a class="dropdown-item" href="../articles/using_tb.html">Using tensorboard</a></li>
    <li><a class="dropdown-item" href="../articles/integrated_gradient.html">Integrated Gradient</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/GenomeNet/deepG/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Data generator for fasta/fastq files and label targets</h1>
      <small class="dont-index">Source: <a href="https://github.com/GenomeNet/deepG/blob/master/R/generator_csv.R" class="external-link"><code>R/generator_csv.R</code></a></small>
      <div class="d-none name"><code>generator_fasta_label_header_csv.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Iterates over folder containing fasta/fastq files and produces encoding of predictor sequences
and target variables. Targets will be read from fasta headers or a separate csv file.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">generator_fasta_label_header_csv</span><span class="op">(</span></span>
<span>  <span class="va">path_corpus</span>,</span>
<span>  format <span class="op">=</span> <span class="st">"fasta"</span>,</span>
<span>  batch_size <span class="op">=</span> <span class="fl">256</span>,</span>
<span>  maxlen <span class="op">=</span> <span class="fl">250</span>,</span>
<span>  max_iter <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  vocabulary <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a"</span>, <span class="st">"c"</span>, <span class="st">"g"</span>, <span class="st">"t"</span><span class="op">)</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  shuffle_file_order <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  step <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1234</span>,</span>
<span>  shuffle_input <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  file_limit <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  path_file_log <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  vocabulary_label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span>, <span class="st">"z"</span><span class="op">)</span>,</span>
<span>  reverse_complement <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ambiguous_nuc <span class="op">=</span> <span class="st">"zero"</span>,</span>
<span>  proportion_per_seq <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  read_data <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  use_quality_score <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  padding <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  skip_amb_nuc <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  max_samples <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  concat_seq <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  added_label_path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  add_input_as_seq <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  target_from_csv <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  target_split <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  file_filter <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  use_coverage <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  proportion_entries <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sample_by_file_size <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  reverse_complement_encoding <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  n_gram <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  n_gram_stride <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  add_noise <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  return_int <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  reshape_xy <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-path-corpus">path_corpus<a class="anchor" aria-label="anchor" href="#arg-path-corpus"></a></dt>
<dd><p>Input directory where fasta files are located or path to single file ending with fasta or fastq
(as specified in format argument). Can also be a list of directories and/or files.</p></dd>


<dt id="arg-format">format<a class="anchor" aria-label="anchor" href="#arg-format"></a></dt>
<dd><p>File format, either <code>"fasta"</code> or <code>"fastq"</code>.</p></dd>


<dt id="arg-batch-size">batch_size<a class="anchor" aria-label="anchor" href="#arg-batch-size"></a></dt>
<dd><p>Number of samples in one batch.</p></dd>


<dt id="arg-maxlen">maxlen<a class="anchor" aria-label="anchor" href="#arg-maxlen"></a></dt>
<dd><p>Length of predictor sequence.</p></dd>


<dt id="arg-max-iter">max_iter<a class="anchor" aria-label="anchor" href="#arg-max-iter"></a></dt>
<dd><p>Stop after <code>max_iter</code> number of iterations failed to produce a new batch.</p></dd>


<dt id="arg-vocabulary">vocabulary<a class="anchor" aria-label="anchor" href="#arg-vocabulary"></a></dt>
<dd><p>Vector of allowed characters. Characters outside vocabulary get encoded as specified in <code>ambiguous_nuc</code>.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Whether to show messages.</p></dd>


<dt id="arg-shuffle-file-order">shuffle_file_order<a class="anchor" aria-label="anchor" href="#arg-shuffle-file-order"></a></dt>
<dd><p>Logical, whether to go through files randomly or sequentially.</p></dd>


<dt id="arg-step">step<a class="anchor" aria-label="anchor" href="#arg-step"></a></dt>
<dd><p>How often to take a sample.</p></dd>


<dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>Sets seed for <code>set.seed</code> function for reproducible results.</p></dd>


<dt id="arg-shuffle-input">shuffle_input<a class="anchor" aria-label="anchor" href="#arg-shuffle-input"></a></dt>
<dd><p>Whether to shuffle entries in every fasta/fastq file before extracting samples.</p></dd>


<dt id="arg-file-limit">file_limit<a class="anchor" aria-label="anchor" href="#arg-file-limit"></a></dt>
<dd><p>Integer or <code>NULL</code>. If integer, use only specified number of randomly sampled files for training. Ignored if greater than number of files in <code>path</code>.</p></dd>


<dt id="arg-path-file-log">path_file_log<a class="anchor" aria-label="anchor" href="#arg-path-file-log"></a></dt>
<dd><p>Write name of files to csv file if path is specified.</p></dd>


<dt id="arg-vocabulary-label">vocabulary_label<a class="anchor" aria-label="anchor" href="#arg-vocabulary-label"></a></dt>
<dd><p>Character vector of possible targets. Targets outside <code>vocabulary_label</code> will get discarded.</p></dd>


<dt id="arg-reverse-complement">reverse_complement<a class="anchor" aria-label="anchor" href="#arg-reverse-complement"></a></dt>
<dd><p>Boolean, for every new file decide randomly to use original data or its reverse complement.</p></dd>


<dt id="arg-ambiguous-nuc">ambiguous_nuc<a class="anchor" aria-label="anchor" href="#arg-ambiguous-nuc"></a></dt>
<dd><p>How to handle nucleotides outside vocabulary, either <code>"zero"</code>, <code>"discard"</code>, <code>"empirical"</code> or <code>"equal"</code>.</p><ul><li><p>If <code>"zero"</code>, input gets encoded as zero vector.</p></li>
<li><p>If <code>"equal"</code>, input is repetition of <code>1/length(vocabulary)</code>.</p></li>
<li><p>If <code>"discard"</code>, samples containing nucleotides outside vocabulary get discarded.</p></li>
<li><p>If <code>"empirical"</code>, use nucleotide distribution of current file.</p></li>
</ul></dd>


<dt id="arg-proportion-per-seq">proportion_per_seq<a class="anchor" aria-label="anchor" href="#arg-proportion-per-seq"></a></dt>
<dd><p>Numerical value between 0 and 1. Proportion of sequence to take samples from (use random subsequence).</p></dd>


<dt id="arg-read-data">read_data<a class="anchor" aria-label="anchor" href="#arg-read-data"></a></dt>
<dd><p>If <code>TRUE</code> the first element of input is a list of length 2, each containing one part of paired read. Maxlen should be 2*length of one read.</p></dd>


<dt id="arg-use-quality-score">use_quality_score<a class="anchor" aria-label="anchor" href="#arg-use-quality-score"></a></dt>
<dd><p>Whether to use fastq quality scores. If TRUE input is not one-hot-encoding but corresponds to probabilities.
For example (0.97, 0.01, 0.01, 0.01) instead of (1, 0, 0, 0).</p></dd>


<dt id="arg-padding">padding<a class="anchor" aria-label="anchor" href="#arg-padding"></a></dt>
<dd><p>Whether to pad sequences too short for one sample with zeros.</p></dd>


<dt id="arg-skip-amb-nuc">skip_amb_nuc<a class="anchor" aria-label="anchor" href="#arg-skip-amb-nuc"></a></dt>
<dd><p>Threshold of ambiguous nucleotides to accept in fasta entry. Complete entry will get discarded otherwise.</p></dd>


<dt id="arg-max-samples">max_samples<a class="anchor" aria-label="anchor" href="#arg-max-samples"></a></dt>
<dd><p>Maximum number of samples to use from one file. If not <code>NULL</code> and file has more than <code>max_samples</code> samples, will randomly choose a
subset of <code>max_samples</code> samples.</p></dd>


<dt id="arg-concat-seq">concat_seq<a class="anchor" aria-label="anchor" href="#arg-concat-seq"></a></dt>
<dd><p>Character string or <code>NULL</code>. If not <code>NULL</code> all entries from file get concatenated to one sequence with <code>concat_seq</code> string between them.
Example: If 1.entry AACC, 2. entry TTTG and <code>concat_seq = "ZZZ"</code> this becomes AACCZZZTTTG.</p></dd>


<dt id="arg-added-label-path">added_label_path<a class="anchor" aria-label="anchor" href="#arg-added-label-path"></a></dt>
<dd><p>Path to file with additional input labels. Should be a csv file with one column named "file". Other columns should correspond to labels.</p></dd>


<dt id="arg-add-input-as-seq">add_input_as_seq<a class="anchor" aria-label="anchor" href="#arg-add-input-as-seq"></a></dt>
<dd><p>Boolean vector specifying for each entry in <code>added_label_path</code> if rows from csv should be encoded as a sequence or used directly.
If a row in your csv file is a sequence this should be <code>TRUE</code>. For example you may want to add another sequence, say ACCGT. Then this would correspond to 1,2,2,3,4 in
csv file (if vocabulary = c("A", "C", "G", "T")).  If <code>add_input_as_seq</code> is <code>TRUE</code>, 12234 gets one-hot encoded, so added input is a 3D tensor.  If <code>add_input_as_seq</code> is
<code>FALSE</code> this will feed network just raw data (a 2D tensor).</p></dd>


<dt id="arg-target-from-csv">target_from_csv<a class="anchor" aria-label="anchor" href="#arg-target-from-csv"></a></dt>
<dd><p>Path to csv file with target mapping. One column should be called "file" and other entries in row are the targets.</p></dd>


<dt id="arg-target-split">target_split<a class="anchor" aria-label="anchor" href="#arg-target-split"></a></dt>
<dd><p>If target gets read from csv file, list of names to divide target tensor into list of tensors.
Example: if csv file has header names <code>"file", "label_1", "label_2", "label_3"</code> and <code>target_split = list(c("label_1", "label_2"), "label_3")</code>,
this will divide target matrix to list of length 2, where the first element contains columns named <code>"label_1"</code> and <code>"label_2"</code> and the
second entry contains the column named <code>"label_3"</code>.</p></dd>


<dt id="arg-file-filter">file_filter<a class="anchor" aria-label="anchor" href="#arg-file-filter"></a></dt>
<dd><p>Vector of file names to use from path_corpus.</p></dd>


<dt id="arg-use-coverage">use_coverage<a class="anchor" aria-label="anchor" href="#arg-use-coverage"></a></dt>
<dd><p>Integer or <code>NULL</code>. If not <code>NULL</code>, use coverage as encoding rather than one-hot encoding and normalize.
Coverage information must be contained in fasta header: there must be a string <code>"cov_n"</code> in the header, where <code>n</code> is some integer.</p></dd>


<dt id="arg-proportion-entries">proportion_entries<a class="anchor" aria-label="anchor" href="#arg-proportion-entries"></a></dt>
<dd><p>Proportion of fasta entries to keep. For example, if fasta file has 50 entries and <code>proportion_entries = 0.1</code>,
will randomly select 5 entries.</p></dd>


<dt id="arg-sample-by-file-size">sample_by_file_size<a class="anchor" aria-label="anchor" href="#arg-sample-by-file-size"></a></dt>
<dd><p>Sample new file weighted by file size (bigger files more likely).</p></dd>


<dt id="arg-reverse-complement-encoding">reverse_complement_encoding<a class="anchor" aria-label="anchor" href="#arg-reverse-complement-encoding"></a></dt>
<dd><p>Whether to use both original sequence and reverse complement as two input sequences.</p></dd>


<dt id="arg-n-gram">n_gram<a class="anchor" aria-label="anchor" href="#arg-n-gram"></a></dt>
<dd><p>Integer, encode target not nucleotide wise but combine n nucleotides at once. For example for <code>n=2, "AA" -&gt;  (1, 0,..., 0),</code>
<code>"AC" -&gt;  (0, 1, 0,..., 0), "TT" -&gt; (0,..., 0, 1)</code>, where the one-hot vectors have length <code>length(vocabulary)^n</code>.</p></dd>


<dt id="arg-n-gram-stride">n_gram_stride<a class="anchor" aria-label="anchor" href="#arg-n-gram-stride"></a></dt>
<dd><p>Step size for n-gram encoding. For AACCGGTT with <code>n_gram = 4</code> and <code>n_gram_stride = 2</code>, generator encodes
<code>(AACC), (CCGG), (GGTT)</code>; for <code>n_gram_stride = 4</code> generator encodes <code>(AACC), (GGTT)</code>.</p></dd>


<dt id="arg-add-noise">add_noise<a class="anchor" aria-label="anchor" href="#arg-add-noise"></a></dt>
<dd><p><code>NULL</code> or list of arguments. If not <code>NULL</code>, list must contain the following arguments: <code>noise_type</code> can be <code>"normal"</code> or <code>"uniform"</code>;
optional arguments <code>sd</code> or <code>mean</code> if noise_type is <code>"normal"</code> (default is <code>sd=1</code> and <code>mean=0</code>) or <code>min, max</code> if <code>noise_type</code> is <code>"uniform"</code>
(default is <code>min=0, max=1</code>).</p></dd>


<dt id="arg-return-int">return_int<a class="anchor" aria-label="anchor" href="#arg-return-int"></a></dt>
<dd><p>Whether to return integer encoding or one-hot encoding.</p></dd>


<dt id="arg-reshape-xy">reshape_xy<a class="anchor" aria-label="anchor" href="#arg-reshape-xy"></a></dt>
<dd><p>Can be a list of functions to apply to input and/or target. List elements (containing the reshape functions)
must be called x for input or y for target and each have arguments called x and y. For example:
<code>reshape_xy = list(x = function(x, y) {return(x+1)}, y = function(x, y) {return(x+y)})</code> .
For rds generator needs to have an additional argument called sw.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A generator function.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># reticulate::py_module_available("tensorflow")</span></span></span>
<span class="r-in"><span><span class="va">path_input</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">path_input</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># create 2 fasta files called 'file_1.fasta', 'file_2.fasta'</span></span></span>
<span class="r-in"><span><span class="fu"><a href="create_dummy_data.html">create_dummy_data</a></span><span class="op">(</span>file_path <span class="op">=</span> <span class="va">path_input</span>, </span></span>
<span class="r-in"><span>                  num_files <span class="op">=</span> <span class="fl">2</span>,</span></span>
<span class="r-in"><span>                  seq_length <span class="op">=</span> <span class="fl">5</span>,</span></span>
<span class="r-in"><span>                  num_seq <span class="op">=</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>                  vocabulary <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a"</span>, <span class="st">"c"</span>, <span class="st">"g"</span>, <span class="st">"t"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">dummy_labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'file_1.fasta'</span>, <span class="st">'file_2.fasta'</span><span class="op">)</span>, <span class="co"># dummy labels</span></span></span>
<span class="r-in"><span>                           label1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                           label2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">target_from_csv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>fileext <span class="op">=</span> <span class="st">'.csv'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.csv</a></span><span class="op">(</span><span class="va">dummy_labels</span>, <span class="va">target_from_csv</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gen</span> <span class="op">&lt;-</span> <span class="fu">generator_fasta_label_header_csv</span><span class="op">(</span>path_corpus <span class="op">=</span> <span class="va">path_input</span>, batch_size <span class="op">=</span> <span class="fl">2</span>, </span></span>
<span class="r-in"><span>                                        maxlen <span class="op">=</span> <span class="fl">5</span>, target_from_csv <span class="op">=</span> <span class="va">target_from_csv</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">z</span> <span class="op">&lt;-</span> <span class="fu">gen</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">z</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">z</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Philipp Münch, René Mreches, Martin Binder, Hüseyin Anil Gündüz, Xiao-Yin To, Alice McHardy.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

