<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index • deepG</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><meta property="og:image" content="https://genomenet.github.io/deepG/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">deepG</a>

    <small class="nav-text text-default me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.3.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html"><span class="fa fa-file-alt"></span> Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-notebooks" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Notebooks</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-notebooks"><li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/175jIdXcDcgPUvaBo2rH2Lupbpjnp5O7G?usp=sharing">deepG tutorial</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/1Eolc0koMNM1zkuO4XyVM58ImeF1BpRiH?usp=sharing">Read-length level: Human contamination</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/1yiXSwFafXpMLHaov9iBTQLIDZ6bK1zYX?usp=sharing">Locus level: CRISPR detection</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/1G7bOFEX87cZNrM2tdRtTdkrZn5fM__g0?usp=sharing">Gene level: 16S rRNA detection</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/1BCggL-tfQF136YeJ8cKKi-zoBEDMgkNh?usp=sharing">Genome level: Bacterial morphology (Sporulation)</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/10xpRzGd3JeBAbqQYSCxzQUMctt01sx9D?usp=sharing">Full metagenome level: Colorectal cancer prediction</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/1kyYK7IU7GSfdpDzO_a8U3_qD4i3zTu6w?usp=sharing">BERT with deepG</a></li>
  </ul></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-tutorials" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Tutorials</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-tutorials"><li><a class="dropdown-item" href="../articles/getting_started.html">Getting Started</a></li>
    <li><a class="dropdown-item" href="../articles/training_types.html">Training types</a></li>
    <li><a class="dropdown-item" href="../articles/data_generator.html">Data generator</a></li>
    <li><a class="dropdown-item" href="../articles/using_tb.html">Using tensorboard</a></li>
    <li><a class="dropdown-item" href="../articles/integrated_gradient.html">Integrated Gradient</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/GenomeNet/deepG/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="training">Training<a class="anchor" aria-label="anchor" href="#training"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="train_model.html">train_model()</a></code>

        </dt>
        <dd>Train neural network on genomic data</dd>
      </dl><dl><dt>

          <code><a href="train_model_cpc.html">train_model_cpc()</a></code>

        </dt>
        <dd>Train CPC inspired model</dd>
      </dl><dl><dt>

          <code><a href="resume_training_from_model_card.html">resume_training_from_model_card()</a></code>

        </dt>
        <dd>Continue training from model card</dd>
      </dl></div><div class="section level2">
      <h2 id="model-architecture">Model Architecture<a class="anchor" aria-label="anchor" href="#model-architecture"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="create_model_genomenet.html">create_model_genomenet()</a></code>

        </dt>
        <dd>Create GenomeNet Model with Given Architecture Parameters</dd>
      </dl><dl><dt>

          <code><a href="create_model_lstm_cnn.html">create_model_lstm_cnn()</a></code>

        </dt>
        <dd>Create LSTM/CNN network</dd>
      </dl><dl><dt>

          <code><a href="create_model_lstm_cnn_multi_input.html">create_model_lstm_cnn_multi_input()</a></code>

        </dt>
        <dd>Create LSTM/CNN network that can process multiple samples for one target</dd>
      </dl><dl><dt>

          <code><a href="create_model_lstm_cnn_target_middle.html">create_model_lstm_cnn_target_middle()</a></code>

        </dt>
        <dd>Create LSTM/CNN network to predict middle part of a sequence</dd>
      </dl><dl><dt>

          <code><a href="create_model_lstm_cnn_time_dist.html">create_model_lstm_cnn_time_dist()</a></code>

        </dt>
        <dd>Create LSTM/CNN network for combining multiple sequences</dd>
      </dl><dl><dt>

          <code><a href="create_model_transformer.html">create_model_transformer()</a></code>

        </dt>
        <dd>Create transformer model</dd>
      </dl><dl><dt>

          <code><a href="create_model_twin_network.html">create_model_twin_network()</a></code>

        </dt>
        <dd>Create twin network</dd>
      </dl><dl><dt>

          <code><a href="remove_add_layers.html">remove_add_layers()</a></code>

        </dt>
        <dd>Remove layers from model and add dense layers</dd>
      </dl><dl><dt>

          <code><a href="merge_models.html">merge_models()</a></code>

        </dt>
        <dd>Merge two models</dd>
      </dl><dl><dt>

          <code><a href="get_output_activations.html">get_output_activations()</a></code>

        </dt>
        <dd>Get activation functions of output layers</dd>
      </dl><dl><dt>

          <code><a href="reshape_input.html">reshape_input()</a></code>

        </dt>
        <dd>Replace input layer</dd>
      </dl><dl><dt>

          <code><a href="load_cp.html">load_cp()</a></code>

        </dt>
        <dd>Load checkpoint</dd>
      </dl><dl><dt>

          <code><a href="compile_model.html">compile_model()</a></code>

        </dt>
        <dd>Compile model</dd>
      </dl><dl><dt>

          <code><a href="layer_aggregate_time_dist_wrapper.html">layer_aggregate_time_dist_wrapper()</a></code>

        </dt>
        <dd>Aggregation layer</dd>
      </dl><dl><dt>

          <code><a href="layer_pos_embedding_wrapper.html">layer_pos_embedding_wrapper()</a></code>

        </dt>
        <dd>Layer for positional embedding</dd>
      </dl><dl><dt>

          <code><a href="layer_pos_sinusoid_wrapper.html">layer_pos_sinusoid_wrapper()</a></code>

        </dt>
        <dd>Layer for positional encoding</dd>
      </dl><dl><dt>

          <code><a href="layer_transformer_block_wrapper.html">layer_transformer_block_wrapper()</a></code>

        </dt>
        <dd>Transformer block</dd>
      </dl></div><div class="section level2">
      <h2 id="callbacks">Callbacks<a class="anchor" aria-label="anchor" href="#callbacks"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="conf_matrix_cb.html">conf_matrix_cb()</a></code>

        </dt>
        <dd>Confusion matrix callback.</dd>
      </dl><dl><dt>

          <code><a href="early_stopping_time_cb.html">early_stopping_time_cb()</a></code>

        </dt>
        <dd>Stop training callback</dd>
      </dl><dl><dt>

          <code><a href="validation_after_training_cb.html">validation_after_training_cb()</a></code>

        </dt>
        <dd>Validation after training callback</dd>
      </dl><dl><dt>

          <code><a href="reset_states_cb.html">reset_states_cb()</a></code>

        </dt>
        <dd>Reset states callback</dd>
      </dl><dl><dt>

          <code><a href="remove_checkpoints.html">remove_checkpoints()</a></code>

        </dt>
        <dd>Remove checkpoints</dd>
      </dl><dl><dt>

          <code><a href="model_card_cb.html">model_card_cb()</a></code>

        </dt>
        <dd>Create model card</dd>
      </dl></div><div class="section level2">
      <h2 id="metrics-and-losses">Metrics and Losses<a class="anchor" aria-label="anchor" href="#metrics-and-losses"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="noisy_loss_wrapper.html">noisy_loss_wrapper()</a></code>

        </dt>
        <dd>Loss function for label noise</dd>
      </dl><dl><dt>

          <code><a href="balanced_acc_wrapper.html">balanced_acc_wrapper()</a></code>

        </dt>
        <dd>Balanced accuracy metric</dd>
      </dl><dl><dt>

          <code><a href="f1_wrapper.html">f1_wrapper()</a></code>

        </dt>
        <dd>F1 metric</dd>
      </dl><dl><dt>

          <code><a href="auc_wrapper.html">auc_wrapper()</a></code>

        </dt>
        <dd>Mean AUC score</dd>
      </dl><dl><dt>

          <code><a href="exp_decay.html">exp_decay()</a></code>

        </dt>
        <dd>Exponential Decay</dd>
      </dl><dl><dt>

          <code><a href="stepdecay.html">stepdecay()</a></code>

        </dt>
        <dd>Step Decay</dd>
      </dl><dl><dt>

          <code><a href="sgdr.html">sgdr()</a></code>

        </dt>
        <dd>Stochastic Gradient Descent with Warm Restarts</dd>
      </dl><dl><dt>

          <code><a href="focal_loss_multiclass.html">focal_loss_multiclass()</a></code>

        </dt>
        <dd>Focal loss for two or more labels</dd>
      </dl><dl><dt>

          <code><a href="loss_cl.html">loss_cl()</a></code>

        </dt>
        <dd>Contrastive loss</dd>
      </dl></div><div class="section level2">
      <h2 id="generators">Generators<a class="anchor" aria-label="anchor" href="#generators"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="generator_dummy.html">generator_dummy()</a></code>

        </dt>
        <dd>Random data generator</dd>
      </dl><dl><dt>

          <code><a href="generator_fasta_label_folder.html">generator_fasta_label_folder()</a></code>

        </dt>
        <dd>Data generator for fasta/fasta files</dd>
      </dl><dl><dt>

          <code><a href="generator_fasta_label_folder_wrapper.html">generator_fasta_label_folder_wrapper()</a></code>

        </dt>
        <dd>Generator wrapper</dd>
      </dl><dl><dt>

          <code><a href="generator_fasta_label_header_csv.html">generator_fasta_label_header_csv()</a></code>

        </dt>
        <dd>Data generator for fasta/fastq files and label targets</dd>
      </dl><dl><dt>

          <code><a href="generator_fasta_lm.html">generator_fasta_lm()</a></code>

        </dt>
        <dd>Language model generator for fasta/fastq files</dd>
      </dl><dl><dt>

          <code><a href="generator_initialize.html">generator_initialize()</a></code>

        </dt>
        <dd>Initializes generators defined by <code>generator_fasta_label_folder</code> function</dd>
      </dl><dl><dt>

          <code><a href="generator_random.html">generator_random()</a></code>

        </dt>
        <dd>Randomly select samples from fasta files</dd>
      </dl><dl><dt>

          <code><a href="generator_rds.html">generator_rds()</a></code>

        </dt>
        <dd>Rds data generator</dd>
      </dl><dl><dt>

          <code><a href="get_generator.html">get_generator()</a></code>

        </dt>
        <dd>Wrapper for generator functions</dd>
      </dl><dl><dt>

          <code><a href="dataset_from_gen.html">dataset_from_gen()</a></code>

        </dt>
        <dd>Collect samples from generator and store in rds or pickle file.</dd>
      </dl></div><div class="section level2">
      <h2 id="prediction">Prediction<a class="anchor" aria-label="anchor" href="#prediction"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="predict_model.html">predict_model()</a></code>

        </dt>
        <dd>Make prediction for nucleotide sequence or entries in fasta/fastq file</dd>
      </dl><dl><dt>

          <code><a href="summarize_states.html">summarize_states()</a></code>

        </dt>
        <dd>Create summary of predictions</dd>
      </dl><dl><dt>

          <code><a href="predict_with_n_gram.html">predict_with_n_gram()</a></code>

        </dt>
        <dd>Predict the next nucleotide using n-gram</dd>
      </dl><dl><dt>

          <code><a href="load_prediction.html">load_prediction()</a></code>

        </dt>
        <dd>Read states from h5 file</dd>
      </dl></div><div class="section level2">
      <h2 id="evaluation">Evaluation<a class="anchor" aria-label="anchor" href="#evaluation"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="evaluate_linear.html">evaluate_linear()</a></code>

        </dt>
        <dd>Evaluate matrices of true targets and predictions from layer with linear activation.</dd>
      </dl><dl><dt>

          <code><a href="evaluate_model.html">evaluate_model()</a></code>

        </dt>
        <dd>Evaluates a trained model on fasta, fastq or rds files</dd>
      </dl><dl><dt>

          <code><a href="evaluate_sigmoid.html">evaluate_sigmoid()</a></code>

        </dt>
        <dd>Evaluate matrices of true targets and predictions from layer with sigmoid activation.</dd>
      </dl><dl><dt>

          <code><a href="evaluate_softmax.html">evaluate_softmax()</a></code>

        </dt>
        <dd>Evaluate matrices of true targets and predictions from layer with softmax activation.</dd>
      </dl><dl><dt>

          <code><a href="plot_roc.html">plot_roc()</a></code>

        </dt>
        <dd>Plot ROC</dd>
      </dl></div><div class="section level2">
      <h2 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="integrated_gradients.html">integrated_gradients()</a></code>

        </dt>
        <dd>Compute integrated gradients</dd>
      </dl><dl><dt>

          <code><a href="heatmaps_integrated_grad.html">heatmaps_integrated_grad()</a></code>

        </dt>
        <dd>Heatmap of integrated gradient scores</dd>
      </dl><dl><dt>

          <code><a href="plot_cm.html">plot_cm()</a></code>

        </dt>
        <dd>Plot confusion matrix</dd>
      </dl></div><div class="section level2">
      <h2 id="preprocessing">Preprocessing<a class="anchor" aria-label="anchor" href="#preprocessing"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="seq_encoding_label.html">seq_encoding_label()</a></code>

        </dt>
        <dd>Encodes integer sequence for label classification.</dd>
      </dl><dl><dt>

          <code><a href="seq_encoding_lm.html">seq_encoding_lm()</a></code>

        </dt>
        <dd>Encodes integer sequence for language model</dd>
      </dl><dl><dt>

          <code><a href="get_class_weight.html">get_class_weight()</a></code>

        </dt>
        <dd>Estimate frequency of different classes</dd>
      </dl><dl><dt>

          <code><a href="get_start_ind.html">get_start_ind()</a></code>

        </dt>
        <dd>Computes start position of samples</dd>
      </dl><dl><dt>

          <code><a href="int_to_n_gram.html">int_to_n_gram()</a></code>

        </dt>
        <dd>Encode sequence of integers to sequence of n-gram</dd>
      </dl><dl><dt>

          <code><a href="n_gram_dist.html">n_gram_dist()</a></code>

        </dt>
        <dd>Get distribution of n-grams</dd>
      </dl><dl><dt>

          <code><a href="n_gram_of_matrix.html">n_gram_of_matrix()</a></code>

        </dt>
        <dd>One-hot encoding matrix to n-gram encoding matrix</dd>
      </dl><dl><dt>

          <code><a href="reshape_tensor.html">reshape_tensor()</a></code>

        </dt>
        <dd>Reshape tensors for set learning</dd>
      </dl><dl><dt>

          <code><a href="split_fasta.html">split_fasta()</a></code>

        </dt>
        <dd>Split fasta file into smaller files.</dd>
      </dl><dl><dt>

          <code><a href="one_hot_to_seq.html">one_hot_to_seq()</a></code>

        </dt>
        <dd>Char sequence corresponding to one-hot matrix.</dd>
      </dl></div><div class="section level2">
      <h2 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="crispr_sample.html">crispr_sample</a></code>

        </dt>
        <dd>CRISPR data</dd>
      </dl><dl><dt>

          <code><a href="ecoli_small.html">ecoli_small</a></code>

        </dt>
        <dd>Ecoli subset</dd>
      </dl><dl><dt>

          <code><a href="parenthesis.html">parenthesis</a></code>

        </dt>
        <dd>Parenthesis data</dd>
      </dl></div><div class="section level2">
      <h2 id="dummy-data">Dummy Data<a class="anchor" aria-label="anchor" href="#dummy-data"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="create_dummy_data.html">create_dummy_data()</a></code>

        </dt>
        <dd>Write random sequences to fasta file</dd>
      </dl></div><div class="section level2">
      <h2 id="session-management">Session Management<a class="anchor" aria-label="anchor" href="#session-management"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="deepG-package.html">deepG</a></code> <code><a href="deepG-package.html">deepG-package</a></code>

        </dt>
        <dd>deepG for GenomeNet</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Philipp Münch, René Mreches, Martin Binder, Hüseyin Anil Gündüz, Xiao-Yin To, Alice McHardy.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

