<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Make prediction for nucleotide sequence or entries in fasta/fastq file — predict_model • deepG</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Make prediction for nucleotide sequence or entries in fasta/fastq file — predict_model"><meta name="description" content='Removes layers (optional) from pretrained model and calculates states of fasta/fastq file or nucleotide sequence.
Writes states to h5 or csv file (access content of h5 output with load_prediction function).
There are several options on how to process an input file:
If "one_seq", computes prediction for sequence argument or fasta/fastq file.
Combines fasta entries in file to one sequence. This means predictor sequences can contain elements from more than one fasta entry.
If "by_entry", will output a separate file for each fasta/fastq entry.
Names of output files are: output_dir + "Nr" + i + filename + output_type, where i is the number of the fasta entry.
If "by_entry_one_file", will store prediction for all fasta entries in one h5 file.
If "one_pred_per_entry", will make one prediction for each entry by either picking random sample for long sequences
or pad sequence for short sequences.

'><meta property="og:description" content='Removes layers (optional) from pretrained model and calculates states of fasta/fastq file or nucleotide sequence.
Writes states to h5 or csv file (access content of h5 output with load_prediction function).
There are several options on how to process an input file:
If "one_seq", computes prediction for sequence argument or fasta/fastq file.
Combines fasta entries in file to one sequence. This means predictor sequences can contain elements from more than one fasta entry.
If "by_entry", will output a separate file for each fasta/fastq entry.
Names of output files are: output_dir + "Nr" + i + filename + output_type, where i is the number of the fasta entry.
If "by_entry_one_file", will store prediction for all fasta entries in one h5 file.
If "one_pred_per_entry", will make one prediction for each entry by either picking random sample for long sequences
or pad sequence for short sequences.

'><meta property="og:image" content="https://genomenet.github.io/deepG/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">deepG</a>

    <small class="nav-text text-default me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.3.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html"><span class="fa fa-file-alt"></span> Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-notebooks" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Notebooks</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-notebooks"><li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/175jIdXcDcgPUvaBo2rH2Lupbpjnp5O7G?usp=sharing">deepG tutorial</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/1Eolc0koMNM1zkuO4XyVM58ImeF1BpRiH?usp=sharing">Read-length level: Human contamination</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/1yiXSwFafXpMLHaov9iBTQLIDZ6bK1zYX?usp=sharing">Locus level: CRISPR detection</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/1G7bOFEX87cZNrM2tdRtTdkrZn5fM__g0?usp=sharing">Gene level: 16S rRNA detection</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/1BCggL-tfQF136YeJ8cKKi-zoBEDMgkNh?usp=sharing">Genome level: Bacterial morphology (Sporulation)</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/10xpRzGd3JeBAbqQYSCxzQUMctt01sx9D?usp=sharing">Full metagenome level: Colorectal cancer prediction</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/1kyYK7IU7GSfdpDzO_a8U3_qD4i3zTu6w?usp=sharing">BERT with deepG</a></li>
  </ul></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-tutorials" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Tutorials</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-tutorials"><li><a class="dropdown-item" href="../articles/getting_started.html">Getting Started</a></li>
    <li><a class="dropdown-item" href="../articles/training_types.html">Training types</a></li>
    <li><a class="dropdown-item" href="../articles/data_generator.html">Data generator</a></li>
    <li><a class="dropdown-item" href="../articles/using_tb.html">Using tensorboard</a></li>
    <li><a class="dropdown-item" href="../articles/integrated_gradient.html">Integrated Gradient</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/GenomeNet/deepG/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Make prediction for nucleotide sequence or entries in fasta/fastq file</h1>
      <small class="dont-index">Source: <a href="https://github.com/GenomeNet/deepG/blob/master/R/predict.R" class="external-link"><code>R/predict.R</code></a></small>
      <div class="d-none name"><code>predict_model.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Removes layers (optional) from pretrained model and calculates states of fasta/fastq file or nucleotide sequence.
Writes states to h5 or csv file (access content of h5 output with <code><a href="load_prediction.html">load_prediction</a></code> function).
There are several options on how to process an input file:</p><ul><li><p>If <code>"one_seq"</code>, computes prediction for sequence argument or fasta/fastq file.
Combines fasta entries in file to one sequence. This means predictor sequences can contain elements from more than one fasta entry.</p></li>
<li><p>If <code>"by_entry"</code>, will output a separate file for each fasta/fastq entry.
Names of output files are: <code>output_dir</code> + "Nr" + i + <code>filename</code> + <code>output_type</code>, where i is the number of the fasta entry.</p></li>
<li><p>If <code>"by_entry_one_file"</code>, will store prediction for all fasta entries in one h5 file.</p></li>
<li><p>If <code>"one_pred_per_entry"</code>, will make one prediction for each entry by either picking random sample for long sequences
or pad sequence for short sequences.</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">predict_model</span><span class="op">(</span></span>
<span>  <span class="va">model</span>,</span>
<span>  output_format <span class="op">=</span> <span class="st">"one_seq"</span>,</span>
<span>  layer_name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sequence <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  path_input <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  round_digits <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  filename <span class="op">=</span> <span class="st">"states.h5"</span>,</span>
<span>  step <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  vocabulary <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a"</span>, <span class="st">"c"</span>, <span class="st">"g"</span>, <span class="st">"t"</span><span class="op">)</span>,</span>
<span>  batch_size <span class="op">=</span> <span class="fl">256</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  return_states <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  output_type <span class="op">=</span> <span class="st">"h5"</span>,</span>
<span>  padding <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  use_quality <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  quality_string <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"label"</span>,</span>
<span>  lm_format <span class="op">=</span> <span class="st">"target_right"</span>,</span>
<span>  output_dir <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  format <span class="op">=</span> <span class="st">"fasta"</span>,</span>
<span>  include_seq <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  reverse_complement_encoding <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  ambiguous_nuc <span class="op">=</span> <span class="st">"zero"</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-model">model<a class="anchor" aria-label="anchor" href="#arg-model"></a></dt>
<dd><p>A keras model.</p></dd>


<dt id="arg-output-format">output_format<a class="anchor" aria-label="anchor" href="#arg-output-format"></a></dt>
<dd><p>Either <code>"one_seq"</code>, <code>"by_entry"</code>, <code>"by_entry_one_file"</code>, <code>"one_pred_per_entry"</code>.</p></dd>


<dt id="arg-layer-name">layer_name<a class="anchor" aria-label="anchor" href="#arg-layer-name"></a></dt>
<dd><p>Name of layer to get output from. If <code>NULL</code>, will use the last layer.</p></dd>


<dt id="arg-sequence">sequence<a class="anchor" aria-label="anchor" href="#arg-sequence"></a></dt>
<dd><p>Character string, ignores path_input if argument given.</p></dd>


<dt id="arg-path-input">path_input<a class="anchor" aria-label="anchor" href="#arg-path-input"></a></dt>
<dd><p>Path to fasta file.</p></dd>


<dt id="arg-round-digits">round_digits<a class="anchor" aria-label="anchor" href="#arg-round-digits"></a></dt>
<dd><p>Number of decimal places.</p></dd>


<dt id="arg-filename">filename<a class="anchor" aria-label="anchor" href="#arg-filename"></a></dt>
<dd><p>Filename to store states in. No file output if argument is <code>NULL</code>.
If <code>output_format = "by_entry"</code>, adds "<em>nr</em>" + "i" after name, where i is entry number.</p></dd>


<dt id="arg-step">step<a class="anchor" aria-label="anchor" href="#arg-step"></a></dt>
<dd><p>Frequency of sampling steps.</p></dd>


<dt id="arg-vocabulary">vocabulary<a class="anchor" aria-label="anchor" href="#arg-vocabulary"></a></dt>
<dd><p>Vector of allowed characters. Characters outside vocabulary get encoded as specified in <code>ambiguous_nuc</code>.</p></dd>


<dt id="arg-batch-size">batch_size<a class="anchor" aria-label="anchor" href="#arg-batch-size"></a></dt>
<dd><p>Number of samples used for one network update.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Boolean.</p></dd>


<dt id="arg-return-states">return_states<a class="anchor" aria-label="anchor" href="#arg-return-states"></a></dt>
<dd><p>Return predictions as data frame. Only supported for output_format <code>"one_seq"</code>.</p></dd>


<dt id="arg-output-type">output_type<a class="anchor" aria-label="anchor" href="#arg-output-type"></a></dt>
<dd><p><code>"h5"</code> or <code>"csv"</code>. If <code>output_format`` is </code>"by_entries_one_file", "one_pred_per_entry"<code>can only be</code>"h5"`.</p></dd>


<dt id="arg-padding">padding<a class="anchor" aria-label="anchor" href="#arg-padding"></a></dt>
<dd><p>Either <code>"none"</code>, <code>"maxlen"</code>, <code>"standard"</code> or <code>"self"</code>.</p><ul><li><p>If <code>"none"</code>, apply no padding and skip sequences that are too short.</p></li>
<li><p>If <code>"maxlen"</code>, pad with maxlen number of zeros vectors.</p></li>
<li><p>If <code>"standard"</code>, pad with zero vectors only if sequence is shorter than maxlen. Pads to minimum size required for one prediction.</p></li>
<li><p>If <code>"self"</code>, concatenate sequence with itself until sequence is long enough for one prediction.
Example: if sequence is "ACGT" and maxlen is 10, make prediction for "ACGTACGTAC".
Only applied if sequence is shorter than maxlen.</p></li>
</ul></dd>


<dt id="arg-use-quality">use_quality<a class="anchor" aria-label="anchor" href="#arg-use-quality"></a></dt>
<dd><p>Whether to use quality scores.</p></dd>


<dt id="arg-quality-string">quality_string<a class="anchor" aria-label="anchor" href="#arg-quality-string"></a></dt>
<dd><p>String for encoding with quality scores (as used in fastq format).</p></dd>


<dt id="arg-mode">mode<a class="anchor" aria-label="anchor" href="#arg-mode"></a></dt>
<dd><p>Either <code>"lm"</code> for language model or <code>"label"</code> for label classification.</p></dd>


<dt id="arg-lm-format">lm_format<a class="anchor" aria-label="anchor" href="#arg-lm-format"></a></dt>
<dd><p>Either <code>"target_right"</code>, <code>"target_middle_lstm"</code>, <code>"target_middle_cnn"</code> or <code>"wavenet"</code>.</p></dd>


<dt id="arg-output-dir">output_dir<a class="anchor" aria-label="anchor" href="#arg-output-dir"></a></dt>
<dd><p>Directory for file output.</p></dd>


<dt id="arg-format">format<a class="anchor" aria-label="anchor" href="#arg-format"></a></dt>
<dd><p>File format, <code>"fasta"</code>, <code>"fastq"</code>, <code>"rds"</code> or <code>"fasta.tar.gz"</code>, <code>"fastq.tar.gz"</code> for <code>tar.gz</code> files.</p></dd>


<dt id="arg-include-seq">include_seq<a class="anchor" aria-label="anchor" href="#arg-include-seq"></a></dt>
<dd><p>Whether to include input sequence in h5 file.</p></dd>


<dt id="arg-reverse-complement-encoding">reverse_complement_encoding<a class="anchor" aria-label="anchor" href="#arg-reverse-complement-encoding"></a></dt>
<dd><p>Whether to use both original sequence and reverse complement as two input sequences.</p></dd>


<dt id="arg-ambiguous-nuc">ambiguous_nuc<a class="anchor" aria-label="anchor" href="#arg-ambiguous-nuc"></a></dt>
<dd><p>How to handle nucleotides outside vocabulary, either <code>"zero"</code>, <code>"discard"</code>, <code>"empirical"</code> or <code>"equal"</code>.</p><ul><li><p>If <code>"zero"</code>, input gets encoded as zero vector.</p></li>
<li><p>If <code>"equal"</code>, input is repetition of <code>1/length(vocabulary)</code>.</p></li>
<li><p>If <code>"discard"</code>, samples containing nucleotides outside vocabulary get discarded.</p></li>
<li><p>If <code>"empirical"</code>, use nucleotide distribution of current file.</p></li>
</ul></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Further arguments for sequence encoding with <code><a href="seq_encoding_label.html">seq_encoding_label</a></code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>If <code>return_states = TRUE</code> returns a list of model predictions and position of corresponding sequences.
If additionally <code>include_seq = TRUE</code>, list contains sequence strings.
If <code>return_states = FALSE</code> returns nothing, just writes output to file(s).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># reticulate::py_module_available("tensorflow")</span></span></span>
<span class="r-in"><span><span class="co"># make prediction for single sequence and write to h5 file</span></span></span>
<span class="r-in"><span><span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="create_model_lstm_cnn.html">create_model_lstm_cnn</a></span><span class="op">(</span>maxlen <span class="op">=</span> <span class="fl">20</span>, layer_lstm <span class="op">=</span> <span class="fl">8</span>, layer_dense <span class="op">=</span> <span class="fl">2</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">vocabulary</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a"</span>, <span class="st">"c"</span>, <span class="st">"g"</span>, <span class="st">"t"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sequence</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">vocabulary</span>, <span class="fl">200</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">output_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>fileext <span class="op">=</span> <span class="st">".h5"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">predict_model</span><span class="op">(</span>output_format <span class="op">=</span> <span class="st">"one_seq"</span>, model <span class="op">=</span> <span class="va">model</span>, step <span class="op">=</span> <span class="fl">10</span>,</span></span>
<span class="r-in"><span>             sequence <span class="op">=</span> <span class="va">sequence</span>, filename <span class="op">=</span> <span class="va">output_file</span>, mode <span class="op">=</span> <span class="st">"label"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># make prediction for fasta file with multiple entries, write output to separate h5 files</span></span></span>
<span class="r-in"><span><span class="va">fasta_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>fileext <span class="op">=</span> <span class="st">".fasta"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="create_dummy_data.html">create_dummy_data</a></span><span class="op">(</span>file_path <span class="op">=</span> <span class="va">fasta_path</span>, num_files <span class="op">=</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>                 num_seq <span class="op">=</span> <span class="fl">5</span>, seq_length <span class="op">=</span> <span class="fl">100</span>,</span></span>
<span class="r-in"><span>                 write_to_file_path <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="create_model_lstm_cnn.html">create_model_lstm_cnn</a></span><span class="op">(</span>maxlen <span class="op">=</span> <span class="fl">20</span>, layer_lstm <span class="op">=</span> <span class="fl">8</span>, layer_dense <span class="op">=</span> <span class="fl">2</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">output_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">output_dir</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">predict_model</span><span class="op">(</span>output_format <span class="op">=</span> <span class="st">"by_entry"</span>, model <span class="op">=</span> <span class="va">model</span>, step <span class="op">=</span> <span class="fl">10</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>               output_dir <span class="op">=</span> <span class="va">output_dir</span>, mode <span class="op">=</span> <span class="st">"label"</span>, path_input <span class="op">=</span> <span class="va">fasta_path</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">output_dir</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Philipp Münch, René Mreches, Martin Binder, Hüseyin Anil Gündüz, Xiao-Yin To, Alice McHardy.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

