<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Train neural network on genomic data — train_model • deepG</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Train neural network on genomic data — train_model"><meta name="description" content="Train a neural network on genomic data. Data can be fasta/fastq files, rds files or a prepared data set.
If the data is given as collection of fasta, fastq or rds files, function will create a data generator that extracts training and validation batches
from files. Function includes several options to determine the sampling strategy of the generator and preprocessing of the data.
Training progress can be visualized in tensorboard. Model weights can be stored during training using checkpoints."><meta property="og:description" content="Train a neural network on genomic data. Data can be fasta/fastq files, rds files or a prepared data set.
If the data is given as collection of fasta, fastq or rds files, function will create a data generator that extracts training and validation batches
from files. Function includes several options to determine the sampling strategy of the generator and preprocessing of the data.
Training progress can be visualized in tensorboard. Model weights can be stored during training using checkpoints."><meta property="og:image" content="https://genomenet.github.io/deepG/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">deepG</a>

    <small class="nav-text text-default me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.3.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html"><span class="fa fa-file-alt"></span> Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-notebooks" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Notebooks</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-notebooks"><li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/175jIdXcDcgPUvaBo2rH2Lupbpjnp5O7G?usp=sharing">deepG tutorial</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/1Eolc0koMNM1zkuO4XyVM58ImeF1BpRiH?usp=sharing">Read-length level: Human contamination</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/1yiXSwFafXpMLHaov9iBTQLIDZ6bK1zYX?usp=sharing">Locus level: CRISPR detection</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/1G7bOFEX87cZNrM2tdRtTdkrZn5fM__g0?usp=sharing">Gene level: 16S rRNA detection</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/1BCggL-tfQF136YeJ8cKKi-zoBEDMgkNh?usp=sharing">Genome level: Bacterial morphology (Sporulation)</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/10xpRzGd3JeBAbqQYSCxzQUMctt01sx9D?usp=sharing">Full metagenome level: Colorectal cancer prediction</a></li>
    <li><a class="external-link dropdown-item" href="https://colab.research.google.com/drive/1kyYK7IU7GSfdpDzO_a8U3_qD4i3zTu6w?usp=sharing">BERT with deepG</a></li>
  </ul></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-tutorials" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Tutorials</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-tutorials"><li><a class="dropdown-item" href="../articles/getting_started.html">Getting Started</a></li>
    <li><a class="dropdown-item" href="../articles/training_types.html">Training types</a></li>
    <li><a class="dropdown-item" href="../articles/data_generator.html">Data generator</a></li>
    <li><a class="dropdown-item" href="../articles/using_tb.html">Using tensorboard</a></li>
    <li><a class="dropdown-item" href="../articles/integrated_gradient.html">Integrated Gradient</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/GenomeNet/deepG/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Train neural network on genomic data</h1>
      <small class="dont-index">Source: <a href="https://github.com/GenomeNet/deepG/blob/master/R/train.R" class="external-link"><code>R/train.R</code></a></small>
      <div class="d-none name"><code>train_model.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Train a neural network on genomic data. Data can be fasta/fastq files, rds files or a prepared data set.
If the data is given as collection of fasta, fastq or rds files, function will create a data generator that extracts training and validation batches
from files. Function includes several options to determine the sampling strategy of the generator and preprocessing of the data.
Training progress can be visualized in tensorboard. Model weights can be stored during training using checkpoints.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">train_model</span><span class="op">(</span></span>
<span>  model <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  dataset <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  dataset_val <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  train_val_ratio <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  run_name <span class="op">=</span> <span class="st">"run_1"</span>,</span>
<span>  initial_epoch <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  class_weight <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  print_scores <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  epochs <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  max_queue_size <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  steps_per_epoch <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  path_checkpoint <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  path_tensorboard <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  path_log <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  save_best_only <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  save_weights_only <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  tb_images <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  path_file_log <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  reset_states <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  early_stopping_time <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  validation_only_after_training <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  train_val_split_csv <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  reduce_lr_on_plateau <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  lr_plateau_factor <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  patience <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  cooldown <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  model_card <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  callback_list <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  train_type <span class="op">=</span> <span class="st">"label_folder"</span>,</span>
<span>  path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  path_val <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  batch_size <span class="op">=</span> <span class="fl">64</span>,</span>
<span>  step <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  shuffle_file_order <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  vocabulary <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a"</span>, <span class="st">"c"</span>, <span class="st">"g"</span>, <span class="st">"t"</span><span class="op">)</span>,</span>
<span>  format <span class="op">=</span> <span class="st">"fasta"</span>,</span>
<span>  ambiguous_nuc <span class="op">=</span> <span class="st">"zero"</span>,</span>
<span>  seed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1234</span>, <span class="fl">4321</span><span class="op">)</span>,</span>
<span>  file_limit <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  use_coverage <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  set_learning <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  proportion_entries <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sample_by_file_size <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  n_gram <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  n_gram_stride <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  masked_lm <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  random_sampling <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  add_noise <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  return_int <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  maxlen <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  reverse_complement <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  reverse_complement_encoding <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  output_format <span class="op">=</span> <span class="st">"target_right"</span>,</span>
<span>  proportion_per_seq <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  read_data <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  use_quality_score <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  padding <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  concat_seq <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  target_len <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  skip_amb_nuc <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  max_samples <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  added_label_path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  add_input_as_seq <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  target_from_csv <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  target_split <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  shuffle_input <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  vocabulary_label <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  delete_used_files <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  reshape_xy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  return_gen <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-model">model<a class="anchor" aria-label="anchor" href="#arg-model"></a></dt>
<dd><p>A keras model.</p></dd>


<dt id="arg-dataset">dataset<a class="anchor" aria-label="anchor" href="#arg-dataset"></a></dt>
<dd><p>List of training data holding training samples in RAM instead of using generator. Should be list with two entries called <code>"X"</code> and <code>"Y"</code>.</p></dd>


<dt id="arg-dataset-val">dataset_val<a class="anchor" aria-label="anchor" href="#arg-dataset-val"></a></dt>
<dd><p>List of validation data. Should have two entries called <code>"X"</code> and <code>"Y"</code>.</p></dd>


<dt id="arg-train-val-ratio">train_val_ratio<a class="anchor" aria-label="anchor" href="#arg-train-val-ratio"></a></dt>
<dd><p>For generator defines the fraction of batches that will be used for validation (compared to size of training data), i.e. one validation iteration
processes <code>batch_size</code> \(*\) <code>steps_per_epoch</code> \(*\) <code>train_val_ratio</code> samples. If you use dataset instead of generator and <code>dataset_val</code> is <code>NULL</code>, splits <code>dataset</code>
into train/validation data.</p></dd>


<dt id="arg-run-name">run_name<a class="anchor" aria-label="anchor" href="#arg-run-name"></a></dt>
<dd><p>Name of the run. Name will be used to identify output from callbacks. If <code>NULL</code>, will use date as run name.
If name already present, will add <code>"_2"</code> to name or <code>"_{x+1}"</code> if name ends with <code>_x</code>, where <code>x</code> is some integer.</p></dd>


<dt id="arg-initial-epoch">initial_epoch<a class="anchor" aria-label="anchor" href="#arg-initial-epoch"></a></dt>
<dd><p>Epoch at which to start training. Note that network
will run for (<code>epochs</code> - <code>initial_epochs</code>) rounds and not <code>epochs</code> rounds.</p></dd>


<dt id="arg-class-weight">class_weight<a class="anchor" aria-label="anchor" href="#arg-class-weight"></a></dt>
<dd><p>List of weights for output. Order should correspond to <code>vocabulary_label</code>.
You can use <code><a href="get_class_weight.html">get_class_weight</a></code> function to estimate class weights:</p>
<p><code>class_weights &lt;- get_class_weights(path = path, train_type = train_type)</code></p>
<p>If <code>train_type = "label_csv"</code> you need to add path to csv file:</p>
<p><code>class_weights &lt;- get_class_weights(path = path, train_type = train_type, csv_path = target_from_csv)</code></p></dd>


<dt id="arg-print-scores">print_scores<a class="anchor" aria-label="anchor" href="#arg-print-scores"></a></dt>
<dd><p>Whether to print train/validation scores during training.</p></dd>


<dt id="arg-epochs">epochs<a class="anchor" aria-label="anchor" href="#arg-epochs"></a></dt>
<dd><p>Number of iterations.</p></dd>


<dt id="arg-max-queue-size">max_queue_size<a class="anchor" aria-label="anchor" href="#arg-max-queue-size"></a></dt>
<dd><p>Maximum size for the generator queue.</p></dd>


<dt id="arg-steps-per-epoch">steps_per_epoch<a class="anchor" aria-label="anchor" href="#arg-steps-per-epoch"></a></dt>
<dd><p>Number of training batches per epoch.</p></dd>


<dt id="arg-path-checkpoint">path_checkpoint<a class="anchor" aria-label="anchor" href="#arg-path-checkpoint"></a></dt>
<dd><p>Path to checkpoints folder or <code>NULL</code>. If <code>NULL</code>, checkpoints don't get stored.</p></dd>


<dt id="arg-path-tensorboard">path_tensorboard<a class="anchor" aria-label="anchor" href="#arg-path-tensorboard"></a></dt>
<dd><p>Path to tensorboard directory or <code>NULL</code>. If <code>NULL</code>, training not tracked on tensorboard.</p></dd>


<dt id="arg-path-log">path_log<a class="anchor" aria-label="anchor" href="#arg-path-log"></a></dt>
<dd><p>Path to directory to write training scores. File name is <code>run_name</code> + <code>".csv"</code>. No output if <code>NULL</code>.</p></dd>


<dt id="arg-save-best-only">save_best_only<a class="anchor" aria-label="anchor" href="#arg-save-best-only"></a></dt>
<dd><p>Only save model that improved on some score. Not applied if argument is <code>NULL</code>. Otherwise must be
list with argument <code>monitor</code> or <code>save_freq</code> (can only use one option). <code>moniter</code> specifies what metric to use.
<code>save_freq</code>, integer specifying how often to store a checkpoint (in epochs).</p></dd>


<dt id="arg-save-weights-only">save_weights_only<a class="anchor" aria-label="anchor" href="#arg-save-weights-only"></a></dt>
<dd><p>Whether to save weights only.</p></dd>


<dt id="arg-tb-images">tb_images<a class="anchor" aria-label="anchor" href="#arg-tb-images"></a></dt>
<dd><p>Whether to show custom images (confusion matrix) in tensorboard "IMAGES" tab.</p></dd>


<dt id="arg-path-file-log">path_file_log<a class="anchor" aria-label="anchor" href="#arg-path-file-log"></a></dt>
<dd><p>Write name of files used for training to csv file if path is specified.</p></dd>


<dt id="arg-reset-states">reset_states<a class="anchor" aria-label="anchor" href="#arg-reset-states"></a></dt>
<dd><p>Whether to reset hidden states of RNN layer at every new input file and before/after validation.</p></dd>


<dt id="arg-early-stopping-time">early_stopping_time<a class="anchor" aria-label="anchor" href="#arg-early-stopping-time"></a></dt>
<dd><p>Time in seconds after which to stop training.</p></dd>


<dt id="arg-validation-only-after-training">validation_only_after_training<a class="anchor" aria-label="anchor" href="#arg-validation-only-after-training"></a></dt>
<dd><p>Whether to skip validation during training and only do one validation iteration after training.</p></dd>


<dt id="arg-train-val-split-csv">train_val_split_csv<a class="anchor" aria-label="anchor" href="#arg-train-val-split-csv"></a></dt>
<dd><p>A csv file specifying train/validation split. csv file should contain one column named <code>"file"</code> and one column named
<code>"type"</code>. The <code>"file"</code> column contains names of fasta/fastq files and <code>"type"</code> column specifies if file is used for training or validation.
Entries in <code>"type"</code> must be named <code>"train"</code> or <code>"val"</code>, otherwise file will not be used for either. <code>path</code> and <code>path_val</code> arguments should be the same.
Not implemented for <code>train_type = "label_folder"</code>.</p></dd>


<dt id="arg-reduce-lr-on-plateau">reduce_lr_on_plateau<a class="anchor" aria-label="anchor" href="#arg-reduce-lr-on-plateau"></a></dt>
<dd><p>Whether to use learning rate scheduler.</p></dd>


<dt id="arg-lr-plateau-factor">lr_plateau_factor<a class="anchor" aria-label="anchor" href="#arg-lr-plateau-factor"></a></dt>
<dd><p>Factor of decreasing learning rate when plateau is reached.</p></dd>


<dt id="arg-patience">patience<a class="anchor" aria-label="anchor" href="#arg-patience"></a></dt>
<dd><p>Number of epochs waiting for decrease in validation loss before reducing learning rate.</p></dd>


<dt id="arg-cooldown">cooldown<a class="anchor" aria-label="anchor" href="#arg-cooldown"></a></dt>
<dd><p>Number of epochs without changing learning rate.</p></dd>


<dt id="arg-model-card">model_card<a class="anchor" aria-label="anchor" href="#arg-model-card"></a></dt>
<dd><p>List of arguments for training parameters of training run. Must contain at least an entry <code>path_model_card</code>, i.e. the
directory where parameters are stored. List can contain additional (optional) arguments, for example
<code>model_card = list(path_model_card = "/path/to/logs", description = "transfer learning with BERT model on virus data", ...)</code></p></dd>


<dt id="arg-callback-list">callback_list<a class="anchor" aria-label="anchor" href="#arg-callback-list"></a></dt>
<dd><p>Add additional callbacks to <code><a href="https://generics.r-lib.org/reference/fit.html" class="external-link">keras::fit</a></code> call.</p></dd>


<dt id="arg-train-type">train_type<a class="anchor" aria-label="anchor" href="#arg-train-type"></a></dt>
<dd><p>Either <code>"lm"</code>, <code>"lm_rds"</code>, <code>"masked_lm"</code> for language model; <code>"label_header"</code>, <code>"label_folder"</code>, <code>"label_csv"</code>, <code>"label_rds"</code> for classification or <code>"dummy_gen"</code>.</p><ul><li><p>Language model is trained to predict character(s) in a sequence. <br></p></li>
<li><p><code>"label_header"</code>/<code>"label_folder"</code>/<code>"label_csv"</code> are trained to predict a corresponding class given a sequence as input.</p></li>
<li><p>If <code>"label_header"</code>, class will be read from fasta headers.</p></li>
<li><p>If <code>"label_folder"</code>, class will be read from folder, i.e. all files in one folder must belong to the same class.</p></li>
<li><p>If <code>"label_csv"</code>, targets are read from a csv file. This file should have one column named "file". The targets then correspond to entries in that row (except "file"
column). Example: if we are currently working with a file called "a.fasta" and corresponding label is "label_1", there should be a row in our csv file</p><table class="table table"><tr><td>file</td><td>label_1</td><td>label_2</td></tr><tr><td>"a.fasta"</td><td>1</td><td>0</td></tr></table></li>
<li><p>If <code>"label_rds"</code>, generator will iterate over set of .rds files containing each a list of input and target tensors. Not implemented for model
with multiple inputs.</p></li>
<li><p>If <code>"lm_rds"</code>, generator will iterate over set of .rds files and will split tensor according to <code>target_len</code> argument
(targets are last <code>target_len</code> nucleotides of each sequence).</p></li>
<li><p>If <code>"dummy_gen"</code>, generator creates random data once and repeatedly feeds these to model.</p></li>
<li><p>If <code>"masked_lm"</code>, generator maskes some parts of the input. See <code>masked_lm</code> argument for details.</p></li>
</ul></dd>


<dt id="arg-path">path<a class="anchor" aria-label="anchor" href="#arg-path"></a></dt>
<dd><p>Path to training data. If <code>train_type</code> is <code>label_folder</code>, should be a vector or list
where each entry corresponds to a class (list elements can be directories and/or individual files). If <code>train_type</code> is not <code>label_folder</code>,
can be a single directory or file or a list of directories and/or files.</p></dd>


<dt id="arg-path-val">path_val<a class="anchor" aria-label="anchor" href="#arg-path-val"></a></dt>
<dd><p>Path to validation data. See <code>path</code> argument for details.</p></dd>


<dt id="arg-batch-size">batch_size<a class="anchor" aria-label="anchor" href="#arg-batch-size"></a></dt>
<dd><p>Number of samples used for one network update.</p></dd>


<dt id="arg-step">step<a class="anchor" aria-label="anchor" href="#arg-step"></a></dt>
<dd><p>Frequency of sampling steps.</p></dd>


<dt id="arg-shuffle-file-order">shuffle_file_order<a class="anchor" aria-label="anchor" href="#arg-shuffle-file-order"></a></dt>
<dd><p>Boolean, whether to go through files sequentially or shuffle beforehand.</p></dd>


<dt id="arg-vocabulary">vocabulary<a class="anchor" aria-label="anchor" href="#arg-vocabulary"></a></dt>
<dd><p>Vector of allowed characters. Characters outside vocabulary get encoded as specified in <code>ambiguous_nuc</code>.</p></dd>


<dt id="arg-format">format<a class="anchor" aria-label="anchor" href="#arg-format"></a></dt>
<dd><p>File format, <code>"fasta"</code>, <code>"fastq"</code>, <code>"rds"</code> or <code>"fasta.tar.gz"</code>, <code>"fastq.tar.gz"</code> for <code>tar.gz</code> files.</p></dd>


<dt id="arg-ambiguous-nuc">ambiguous_nuc<a class="anchor" aria-label="anchor" href="#arg-ambiguous-nuc"></a></dt>
<dd><p>How to handle nucleotides outside vocabulary, either <code>"zero"</code>, <code>"discard"</code>, <code>"empirical"</code> or <code>"equal"</code>.</p><ul><li><p>If <code>"zero"</code>, input gets encoded as zero vector.</p></li>
<li><p>If <code>"equal"</code>, input is repetition of <code>1/length(vocabulary)</code>.</p></li>
<li><p>If <code>"discard"</code>, samples containing nucleotides outside vocabulary get discarded.</p></li>
<li><p>If <code>"empirical"</code>, use nucleotide distribution of current file.</p></li>
</ul></dd>


<dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>Sets seed for reproducible results.</p></dd>


<dt id="arg-file-limit">file_limit<a class="anchor" aria-label="anchor" href="#arg-file-limit"></a></dt>
<dd><p>Integer or <code>NULL</code>. If integer, use only specified number of randomly sampled files for training. Ignored if greater than number of files in <code>path</code>.</p></dd>


<dt id="arg-use-coverage">use_coverage<a class="anchor" aria-label="anchor" href="#arg-use-coverage"></a></dt>
<dd><p>Integer or <code>NULL</code>. If not <code>NULL</code>, use coverage as encoding rather than one-hot encoding and normalize.
Coverage information must be contained in fasta header: there must be a string <code>"cov_n"</code> in the header, where <code>n</code> is some integer.</p></dd>


<dt id="arg-set-learning">set_learning<a class="anchor" aria-label="anchor" href="#arg-set-learning"></a></dt>
<dd><p>When you want to assign one label to set of samples. Only implemented for <code>train_type = "label_folder"</code>.
Input is a list with the following parameters</p><ul><li><p><code>samples_per_target</code>: how many samples to use for one target.</p></li>
<li><p><code>maxlen</code>: length of one sample.</p></li>
<li><p><code>reshape_mode</code>: <code>"time_dist", "multi_input"</code> or <code>"concat"</code>.</p><ul><li><p>If <code>reshape_mode</code> is <code>"multi_input"</code>, generator will produce <code>samples_per_target</code> separate inputs, each of length <code>maxlen</code> (model should have
<code>samples_per_target</code> input layers).</p></li>
<li><p>If reshape_mode is <code>"time_dist"</code>, generator will produce a 4D input array. The dimensions correspond to
<code>(batch_size, samples_per_target, maxlen, length(vocabulary))</code>.</p></li>
<li><p>If <code>reshape_mode</code> is <code>"concat"</code>, generator will concatenate <code>samples_per_target</code> sequences
of length <code>maxlen</code> to one long sequence.</p></li>
</ul></li>
<li><p>If <code>reshape_mode</code> is <code>"concat"</code>, there is an additional <code>buffer_len</code>
argument. If <code>buffer_len</code> is an integer, the subsequences are interspaced with <code>buffer_len</code> rows. The input length is
(<code>maxlen</code> \(*\) <code>samples_per_target</code>) + <code>buffer_len</code> \(*\) (<code>samples_per_target</code> - 1).</p></li>
</ul></dd>


<dt id="arg-proportion-entries">proportion_entries<a class="anchor" aria-label="anchor" href="#arg-proportion-entries"></a></dt>
<dd><p>Proportion of fasta entries to keep. For example, if fasta file has 50 entries and <code>proportion_entries = 0.1</code>,
will randomly select 5 entries.</p></dd>


<dt id="arg-sample-by-file-size">sample_by_file_size<a class="anchor" aria-label="anchor" href="#arg-sample-by-file-size"></a></dt>
<dd><p>Sample new file weighted by file size (bigger files more likely).</p></dd>


<dt id="arg-n-gram">n_gram<a class="anchor" aria-label="anchor" href="#arg-n-gram"></a></dt>
<dd><p>Integer, encode target not nucleotide wise but combine n nucleotides at once. For example for <code>n=2, "AA" -&gt;  (1, 0,..., 0),</code>
<code>"AC" -&gt;  (0, 1, 0,..., 0), "TT" -&gt; (0,..., 0, 1)</code>, where the one-hot vectors have length <code>length(vocabulary)^n</code>.</p></dd>


<dt id="arg-n-gram-stride">n_gram_stride<a class="anchor" aria-label="anchor" href="#arg-n-gram-stride"></a></dt>
<dd><p>Step size for n-gram encoding. For AACCGGTT with <code>n_gram = 4</code> and <code>n_gram_stride = 2</code>, generator encodes
<code>(AACC), (CCGG), (GGTT)</code>; for <code>n_gram_stride = 4</code> generator encodes <code>(AACC), (GGTT)</code>.</p></dd>


<dt id="arg-masked-lm">masked_lm<a class="anchor" aria-label="anchor" href="#arg-masked-lm"></a></dt>
<dd><p>If not <code>NULL</code>, input and target are equal except some parts of the input are masked or random.
Must be list with the following arguments:</p><ul><li><p><code>mask_rate</code>: Rate of input to mask (rate of input to replace with mask token).</p></li>
<li><p><code>random_rate</code>: Rate of input to set to random token.</p></li>
<li><p><code>identity_rate</code>: Rate of input where sample weights are applied but input and output are identical.</p></li>
<li><p><code>include_sw</code>: Whether to include sample weights.</p></li>
<li><p><code>block_len</code> (optional): Masked/random/identity regions appear in blocks of size <code>block_len</code>.</p></li>
</ul></dd>


<dt id="arg-random-sampling">random_sampling<a class="anchor" aria-label="anchor" href="#arg-random-sampling"></a></dt>
<dd><p>Whether samples should be taken from random positions when using <code>max_samples</code> argument. If <code>FALSE</code> random
samples are taken from a consecutive subsequence.</p></dd>


<dt id="arg-add-noise">add_noise<a class="anchor" aria-label="anchor" href="#arg-add-noise"></a></dt>
<dd><p><code>NULL</code> or list of arguments. If not <code>NULL</code>, list must contain the following arguments: <code>noise_type</code> can be <code>"normal"</code> or <code>"uniform"</code>;
optional arguments <code>sd</code> or <code>mean</code> if noise_type is <code>"normal"</code> (default is <code>sd=1</code> and <code>mean=0</code>) or <code>min, max</code> if <code>noise_type</code> is <code>"uniform"</code>
(default is <code>min=0, max=1</code>).</p></dd>


<dt id="arg-return-int">return_int<a class="anchor" aria-label="anchor" href="#arg-return-int"></a></dt>
<dd><p>Whether to return integer encoding or one-hot encoding.</p></dd>


<dt id="arg-maxlen">maxlen<a class="anchor" aria-label="anchor" href="#arg-maxlen"></a></dt>
<dd><p>Length of predictor sequence.</p></dd>


<dt id="arg-reverse-complement">reverse_complement<a class="anchor" aria-label="anchor" href="#arg-reverse-complement"></a></dt>
<dd><p>Boolean, for every new file decide randomly to use original data or its reverse complement.</p></dd>


<dt id="arg-reverse-complement-encoding">reverse_complement_encoding<a class="anchor" aria-label="anchor" href="#arg-reverse-complement-encoding"></a></dt>
<dd><p>Whether to use both original sequence and reverse complement as two input sequences.</p></dd>


<dt id="arg-output-format">output_format<a class="anchor" aria-label="anchor" href="#arg-output-format"></a></dt>
<dd><p>Determines shape of output tensor for language model.
Either <code>"target_right"</code>, <code>"target_middle_lstm"</code>, <code>"target_middle_cnn"</code> or <code>"wavenet"</code>.
Assume a sequence <code>"AACCGTA"</code>. Output correspond as follows</p><ul><li><p><code>"target_right": X = "AACCGT", Y = "A"</code></p></li>
<li><p><code>"target_middle_lstm": X = (X_1 = "AAC", X_2 = "ATG"), Y = "C"</code> (note reversed order of X_2)</p></li>
<li><p><code>"target_middle_cnn": X = "AACGTA", Y = "C"</code></p></li>
<li><p><code>"wavenet": X = "AACCGT", Y = "ACCGTA"</code></p></li>
</ul></dd>


<dt id="arg-proportion-per-seq">proportion_per_seq<a class="anchor" aria-label="anchor" href="#arg-proportion-per-seq"></a></dt>
<dd><p>Numerical value between 0 and 1. Proportion of sequence to take samples from (use random subsequence).</p></dd>


<dt id="arg-read-data">read_data<a class="anchor" aria-label="anchor" href="#arg-read-data"></a></dt>
<dd><p>If <code>TRUE</code> the first element of output is a list of length 2, each containing one part of paired read. Maxlen should be 2*length of one read.</p></dd>


<dt id="arg-use-quality-score">use_quality_score<a class="anchor" aria-label="anchor" href="#arg-use-quality-score"></a></dt>
<dd><p>Whether to use fastq quality scores. If <code>TRUE</code> input is not one-hot-encoding but corresponds to probabilities.
For example (0.97, 0.01, 0.01, 0.01) instead of (1, 0, 0, 0).</p></dd>


<dt id="arg-padding">padding<a class="anchor" aria-label="anchor" href="#arg-padding"></a></dt>
<dd><p>Whether to pad sequences too short for one sample with zeros.</p></dd>


<dt id="arg-concat-seq">concat_seq<a class="anchor" aria-label="anchor" href="#arg-concat-seq"></a></dt>
<dd><p>Character string or <code>NULL</code>. If not <code>NULL</code> all entries from file get concatenated to one sequence with <code>concat_seq</code> string between them.
Example: If 1.entry AACC, 2. entry TTTG and <code>concat_seq = "ZZZ"</code> this becomes AACCZZZTTTG.</p></dd>


<dt id="arg-target-len">target_len<a class="anchor" aria-label="anchor" href="#arg-target-len"></a></dt>
<dd><p>Number of nucleotides to predict at once for language model.</p></dd>


<dt id="arg-skip-amb-nuc">skip_amb_nuc<a class="anchor" aria-label="anchor" href="#arg-skip-amb-nuc"></a></dt>
<dd><p>Threshold of ambiguous nucleotides to accept in fasta entry. Complete entry will get discarded otherwise.</p></dd>


<dt id="arg-max-samples">max_samples<a class="anchor" aria-label="anchor" href="#arg-max-samples"></a></dt>
<dd><p>Maximum number of samples to use from one file. If not <code>NULL</code> and file has more than <code>max_samples</code> samples, will randomly choose a
subset of <code>max_samples</code> samples.</p></dd>


<dt id="arg-added-label-path">added_label_path<a class="anchor" aria-label="anchor" href="#arg-added-label-path"></a></dt>
<dd><p>Path to file with additional input labels. Should be a csv file with one column named "file". Other columns should correspond to labels.</p></dd>


<dt id="arg-add-input-as-seq">add_input_as_seq<a class="anchor" aria-label="anchor" href="#arg-add-input-as-seq"></a></dt>
<dd><p>Boolean vector specifying for each entry in <code>added_label_path</code> if rows from csv should be encoded as a sequence or used directly.
If a row in your csv file is a sequence this should be <code>TRUE</code>. For example you may want to add another sequence, say ACCGT. Then this would correspond to 1,2,2,3,4 in
csv file (if vocabulary = c("A", "C", "G", "T")).  If <code>add_input_as_seq</code> is <code>TRUE</code>, 12234 gets one-hot encoded, so added input is a 3D tensor.  If <code>add_input_as_seq</code> is
<code>FALSE</code> this will feed network just raw data (a 2D tensor).</p></dd>


<dt id="arg-target-from-csv">target_from_csv<a class="anchor" aria-label="anchor" href="#arg-target-from-csv"></a></dt>
<dd><p>Path to csv file with target mapping. One column should be called "file" and other entries in row are the targets.</p></dd>


<dt id="arg-target-split">target_split<a class="anchor" aria-label="anchor" href="#arg-target-split"></a></dt>
<dd><p>If target gets read from csv file, list of names to divide target tensor into list of tensors.
Example: if csv file has header names <code>"file", "label_1", "label_2", "label_3"</code> and <code>target_split = list(c("label_1", "label_2"), "label_3")</code>,
this will divide target matrix to list of length 2, where the first element contains columns named <code>"label_1"</code> and <code>"label_2"</code> and the
second entry contains the column named <code>"label_3"</code>.</p></dd>


<dt id="arg-shuffle-input">shuffle_input<a class="anchor" aria-label="anchor" href="#arg-shuffle-input"></a></dt>
<dd><p>Whether to shuffle entries in file.</p></dd>


<dt id="arg-vocabulary-label">vocabulary_label<a class="anchor" aria-label="anchor" href="#arg-vocabulary-label"></a></dt>
<dd><p>Character vector of possible targets. Targets outside <code>vocabulary_label</code> will get discarded if
<code>train_type = "label_header"</code>.</p></dd>


<dt id="arg-delete-used-files">delete_used_files<a class="anchor" aria-label="anchor" href="#arg-delete-used-files"></a></dt>
<dd><p>Whether to delete file once used. Only applies for rds files.</p></dd>


<dt id="arg-reshape-xy">reshape_xy<a class="anchor" aria-label="anchor" href="#arg-reshape-xy"></a></dt>
<dd><p>Can be a list of functions to apply to input and/or target. List elements (containing the reshape functions)
must be called x for input or y for target and each have arguments called x and y. For example:
<code>reshape_xy = list(x = function(x, y) {return(x+1)}, y = function(x, y) {return(x+y)})</code> .
For rds generator needs to have an additional argument called sw.</p></dd>


<dt id="arg-return-gen">return_gen<a class="anchor" aria-label="anchor" href="#arg-return-gen"></a></dt>
<dd><p>Whether to return the train and validation generators (instead of training).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list of training metrics.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># reticulate::py_module_available("tensorflow")</span></span></span>
<span class="r-in"><span><span class="co"># create dummy data</span></span></span>
<span class="r-in"><span><span class="va">path_train_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">path_train_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">path_val_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">path_val_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw">for</span> <span class="op">(</span><span class="va">current_path</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">path_train_1</span>, <span class="va">path_train_2</span>,</span></span>
<span class="r-in"><span>                       <span class="va">path_val_1</span>, <span class="va">path_val_2</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">current_path</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="create_dummy_data.html">create_dummy_data</a></span><span class="op">(</span>file_path <span class="op">=</span> <span class="va">current_path</span>,</span></span>
<span class="r-in"><span>                    num_files <span class="op">=</span> <span class="fl">3</span>,</span></span>
<span class="r-in"><span>                    seq_length <span class="op">=</span> <span class="fl">10</span>,</span></span>
<span class="r-in"><span>                    num_seq <span class="op">=</span> <span class="fl">5</span>,</span></span>
<span class="r-in"><span>                    vocabulary <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a"</span>, <span class="st">"c"</span>, <span class="st">"g"</span>, <span class="st">"t"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># create model</span></span></span>
<span class="r-in"><span><span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="create_model_lstm_cnn.html">create_model_lstm_cnn</a></span><span class="op">(</span>layer_lstm <span class="op">=</span> <span class="fl">8</span>, layer_dense <span class="op">=</span> <span class="fl">2</span>, maxlen <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># train model</span></span></span>
<span class="r-in"><span><span class="va">hist</span> <span class="op">&lt;-</span> <span class="fu">train_model</span><span class="op">(</span>train_type <span class="op">=</span> <span class="st">"label_folder"</span>,</span></span>
<span class="r-in"><span>                    model <span class="op">=</span> <span class="va">model</span>,</span></span>
<span class="r-in"><span>                    path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">path_train_1</span>, <span class="va">path_train_2</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                    path_val <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">path_val_1</span>, <span class="va">path_val_2</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                    batch_size <span class="op">=</span> <span class="fl">8</span>,</span></span>
<span class="r-in"><span>                    epochs <span class="op">=</span> <span class="fl">3</span>,</span></span>
<span class="r-in"><span>                    steps_per_epoch <span class="op">=</span> <span class="fl">6</span>,</span></span>
<span class="r-in"><span>                    step <span class="op">=</span> <span class="fl">5</span>,</span></span>
<span class="r-in"><span>                    format <span class="op">=</span> <span class="st">"fasta"</span>,</span></span>
<span class="r-in"><span>                    vocabulary_label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"label_1"</span>, <span class="st">"label_2"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span> </span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Philipp Münch, René Mreches, Martin Binder, Hüseyin Anil Gündüz, Xiao-Yin To, Alice McHardy.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

