<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Language model generator for fasta/fastq files — generator_fasta_lm • deepG</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Language model generator for fasta/fastq files — generator_fasta_lm"><meta property="og:description" content="Iterates over folder containing .fasta/.fastq files and produces encoding of predictor sequences
and target variables. Will take a sequence of fixed size and use some part of sequence as input and other part as target."><meta property="og:image" content="https://genomenet.github.io/deepG/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">deepG</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.1-9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">
    <span class="fa fa-file-alt"></span>
     
    Reference
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Google Colab Notebooks
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="https://colab.research.google.com/drive/175jIdXcDcgPUvaBo2rH2Lupbpjnp5O7G?usp=sharing" class="external-link">deepG tutorial</a>
    </li>
    <li>
      <a href="https://colab.research.google.com/drive/1Eolc0koMNM1zkuO4XyVM58ImeF1BpRiH?usp=sharing" class="external-link">Human contamination</a>
    </li>
    <li>
      <a href="https://colab.research.google.com/drive/1MmQB2pFfEcp4GbgbT7wu6ScNt1YsChmn?usp=sharing" class="external-link">CRISPR detection</a>
    </li>
    <li>
      <a href="https://colab.research.google.com/drive/1G7bOFEX87cZNrM2tdRtTdkrZn5fM__g0?usp=sharing" class="external-link">16S rRNA detection</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/getting_started.html">Getting Started</a>
    </li>
    <li>
      <a href="../articles/training_types.html">Training types</a>
    </li>
    <li>
      <a href="../articles/data_generator.html">Data generator</a>
    </li>
    <li>
      <a href="../articles/using_tb.html">Using tensorboard</a>
    </li>
    <li>
      <a href="../articles/integrated_gradient.html">Integrated Gradient</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/GenomeNet/deepG/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Language model generator for fasta/fastq files</h1>
    <small class="dont-index">Source: <a href="https://github.com/GenomeNet/deepG/blob/HEAD/R/generators.R" class="external-link"><code>R/generators.R</code></a></small>
    <div class="hidden name"><code>generator_fasta_lm.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Iterates over folder containing .fasta/.fastq files and produces encoding of predictor sequences
and target variables. Will take a sequence of fixed size and use some part of sequence as input and other part as target.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">generator_fasta_lm</span><span class="op">(</span></span>
<span>  <span class="va">path_corpus</span>,</span>
<span>  format <span class="op">=</span> <span class="st">"fasta"</span>,</span>
<span>  batch_size <span class="op">=</span> <span class="fl">256</span>,</span>
<span>  maxlen <span class="op">=</span> <span class="fl">250</span>,</span>
<span>  max_iter <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  vocabulary <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a"</span>, <span class="st">"c"</span>, <span class="st">"g"</span>, <span class="st">"t"</span><span class="op">)</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  shuffle_file_order <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  step <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1234</span>,</span>
<span>  shuffle_input <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  file_limit <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  path_file_log <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  reverse_complement <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  output_format <span class="op">=</span> <span class="st">"target_right"</span>,</span>
<span>  ambiguous_nuc <span class="op">=</span> <span class="st">"zeros"</span>,</span>
<span>  use_quality_score <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  proportion_per_seq <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  padding <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  added_label_path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  add_input_as_seq <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  skip_amb_nuc <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  max_samples <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  concat_seq <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  target_len <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  file_filter <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  use_coverage <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  proportion_entries <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sample_by_file_size <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  n_gram <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  n_gram_stride <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  add_noise <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>path_corpus</dt>
<dd><p>Input directory where .fasta files are located or path to single file ending with .fasta or .fastq
(as specified in format argument). Can also be a list of directories and/or files.</p></dd>


<dt>format</dt>
<dd><p>File format, either <code>"fasta"</code> or <code>"fastq"</code>.</p></dd>


<dt>batch_size</dt>
<dd><p>Number of samples in one batch.</p></dd>


<dt>maxlen</dt>
<dd><p>Length of predictor sequence.</p></dd>


<dt>max_iter</dt>
<dd><p>Stop after max_iter number of iterations failed to produce a new batch.</p></dd>


<dt>vocabulary</dt>
<dd><p>Vector of allowed characters, character outside vocabulary get encoded as 0-vector.</p></dd>


<dt>verbose</dt>
<dd><p>Whether to show messages.</p></dd>


<dt>shuffle_file_order</dt>
<dd><p>Logical, whether to go through files randomly or sequentially.</p></dd>


<dt>step</dt>
<dd><p>How often to take a sample.</p></dd>


<dt>seed</dt>
<dd><p>Sets seed for `set.seed`` function for reproducible results.</p></dd>


<dt>shuffle_input</dt>
<dd><p>Whether to shuffle entries in every fasta/fastq file before extracting samples.</p></dd>


<dt>file_limit</dt>
<dd><p>Use only specified number of files, ignored if greater than number of files in path_corpus.</p></dd>


<dt>path_file_log</dt>
<dd><p>Write name of files to csv file if path is specified.</p></dd>


<dt>reverse_complement</dt>
<dd><p>Logical, for every new file decide randomly to use original data or its reverse complement.</p></dd>


<dt>output_format</dt>
<dd><p>Determines shape of output tensor for language model.
Either <code>"target_right"</code>, <code>"target_middle_lstm"</code>, <code>"target_middle_cnn"</code> or <code>"wavenet"</code>.
Assume a sequence <code>"AACCGTA"</code>. Output correspond as follows</p><ul><li><p><code>"target_right": X = "AACCGT", Y = "A"</code></p></li>
<li><p><code>"target_middle_lstm": X = (X_1 = "AAC", X_2 = "ATG"), Y = "C"</code> (note reversed order of X_2)</p></li>
<li><p><code>"target_middle_cnn": X = "AACGTA", Y = "C"</code></p></li>
<li><p><code>"wavenet": X = "AACCGT", Y = "ACCGTA"</code></p></li>
</ul></dd>


<dt>ambiguous_nuc</dt>
<dd><p>How to handle nucleotides outside vocabulary, either <code>"zero"</code>, <code>"discard"</code>, <code>"empirical"</code> or <code>"equal"</code>.</p><ul><li><p>If <code>"zero"</code>, input gets encoded as zero vector.</p></li>
<li><p>If <code>"equal"</code>, input is <code>1/length(vocabulary)</code> \(*\) <code>length(vocabulary)</code>.</p></li>
<li><p>If <code>"discard"</code>, samples containing nucleotides outside vocabulary get discarded.</p></li>
<li><p>If <code>"empirical"</code>, use nucleotide distribution of current file.</p></li>
</ul></dd>


<dt>use_quality_score</dt>
<dd><p>Whether to use fastq quality scores. If TRUE input is not one-hot-encoding but corresponds to probabilities.
For example (0.97, 0.01, 0.01, 0.01) instead of (1, 0, 0, 0).</p></dd>


<dt>proportion_per_seq</dt>
<dd><p>Numerical value between 0 and 1. Proportion of possible samples to take from one file. Takes samples from random subsequence.</p></dd>


<dt>padding</dt>
<dd><p>Whether to pad sequences too short for one sample with zeros.</p></dd>


<dt>added_label_path</dt>
<dd><p>Path to folder with additional input labels. Should be a csv file with one column named "file". Other columns should correspond to one label.</p></dd>


<dt>add_input_as_seq</dt>
<dd><p>Boolean vector specifying for each entry in <code>added_label_path</code> if rows from csv should be encoded as a sequence or used directly.
If a row in your csv file is a sequence this should be true. For example you may want to add another sequence, say ACCGT. Then this would correspond to 1,2,2,3,4 in
csv file (if vocabulary = c("A", "C", "G", "T")).  If <code>add_input_as_seq</code> is TRUE, 12234 gets one-hot encoded, so added input is a 3D tensor.  If <code>add_input_as_seq</code> is
FALSE this will feed network just raw data (a 2D tensor).</p></dd>


<dt>skip_amb_nuc</dt>
<dd><p>Threshold of ambiguous nucleotides to accept in fasta entry. Complete entry will get discarded otherwise.</p></dd>


<dt>max_samples</dt>
<dd><p>Maximum number of samples to use from one file. If not <code>NULL</code> and file has more than <code>max_samples</code> samples, will randomly choose a
subset of <code>max_samples</code> samples.</p></dd>


<dt>concat_seq</dt>
<dd><p>Character string or <code>NULL</code>. If not <code>NULL</code> all entries from file get concatenated to one sequence with concat_seq string between them.
Example: If 1.entry AACC, 2. entry TTTG and <code>concat_seq = "ZZZ"</code> this becomes AACCZZZTTTG.</p></dd>


<dt>target_len</dt>
<dd><p>Number of nucleotides to predict at once.</p></dd>


<dt>file_filter</dt>
<dd><p>Vector of file names to use from path_corpus.</p></dd>


<dt>use_coverage</dt>
<dd><p>Integer or <code>NULL</code>. If not <code>NULL</code>, use coverage as encoding rather than one-hot encoding and normalize.
Coverage information must be contained a fasta header: there must be a string <code>"cov_n"</code> in the header, where <code>n</code> is some integer.</p></dd>


<dt>proportion_entries</dt>
<dd><p>Proportion of fasta entries to keep. For example, if fasta file has 50 entries and proportion_entries = 0.1,
will randomly select 5 entries.</p></dd>


<dt>sample_by_file_size</dt>
<dd><p>Sample new file weighted by file size (bigger files more likely).</p></dd>


<dt>n_gram</dt>
<dd><p>Integer, encode target not nucleotide wise but combine n nucleotides at once. For example for <code>n=2, "AA" -&gt;  (1, 0,..., 0),</code>
<code>"AC" -&gt;  (0, 1, 0,..., 0), "TT" -&gt; (0,..., 0, 1)</code>, where the one-hot vectors have length <code>length(vocabulary)^n</code>.</p></dd>


<dt>n_gram_stride</dt>
<dd><p>Step size for n-gram encoding. For AACCGGTT with <code>n-gram = 4</code> and <code>n_gram_stride = 2</code>, generator encodes
<code>(AACC), (CCGG), (GGTT)</code>; for <code>n_gram_stride = 4</code> generator encodes <code>(AACC), (GGTT)</code>.</p></dd>


<dt>add_noise</dt>
<dd><p><code>NULL</code> or list of arguments. If not <code>NULL</code>, list must contain the following arguments: <code>noise_type</code> can be <code>"normal"</code> or <code>"uniform"</code>;
optional arguments <code>sd</code> or <code>mean</code> if noise_type is <code>"normal"</code> (default is <code>sd=1</code> and <code>mean=0</code>) or <code>min, max</code> if noise_type is <code>"uniform"</code>
(default is <code>min=0, max=1</code>).</p></dd>

</dl></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Philipp Münch, René Mreches, Martin Binder, Hüseyin Anil Gündüz, Alice McHardy.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer></div>

  


  

  </body></html>

