<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Wrapper for generator functions — get_generator • deepG</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Wrapper for generator functions — get_generator"><meta property="og:description" content="For a detailed description see the data generator tutorial.
Will choose one of the generators from generator_fasta_lm,
generator_fasta_label_folder, generator_fasta_label_header_csv,
generator_rds, generator_random, generator_dummy or
generator_fasta_lm according to the train_type and random_sampling
arguments."><meta property="og:image" content="https://genomenet.github.io/deepG/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">deepG</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.1-9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">
    <span class="fa fa-file-alt"></span>
     
    Reference
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Google Colab Notebooks
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="https://colab.research.google.com/drive/175jIdXcDcgPUvaBo2rH2Lupbpjnp5O7G?usp=sharing" class="external-link">deepG tutorial</a>
    </li>
    <li>
      <a href="https://colab.research.google.com/drive/1Eolc0koMNM1zkuO4XyVM58ImeF1BpRiH?usp=sharing" class="external-link">Human contamination</a>
    </li>
    <li>
      <a href="https://colab.research.google.com/drive/1MmQB2pFfEcp4GbgbT7wu6ScNt1YsChmn?usp=sharing" class="external-link">CRISPR detection</a>
    </li>
    <li>
      <a href="https://colab.research.google.com/drive/1G7bOFEX87cZNrM2tdRtTdkrZn5fM__g0?usp=sharing" class="external-link">16S rRNA detection</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/getting_started.html">Getting Started</a>
    </li>
    <li>
      <a href="../articles/training_types.html">Training types</a>
    </li>
    <li>
      <a href="../articles/data_generator.html">Data generator</a>
    </li>
    <li>
      <a href="../articles/using_tb.html">Using tensorboard</a>
    </li>
    <li>
      <a href="../articles/integrated_gradient.html">Integrated Gradient</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/GenomeNet/deepG/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Wrapper for generator functions</h1>
    <small class="dont-index">Source: <a href="https://github.com/GenomeNet/deepG/blob/HEAD/R/generators.R" class="external-link"><code>R/generators.R</code></a></small>
    <div class="hidden name"><code>get_generator.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>For a detailed description see the data generator <a href="https://deepg.de/articles/data_generator.html" class="external-link">tutorial</a>.
Will choose one of the generators from <code><a href="generator_fasta_lm.html">generator_fasta_lm</a></code>,
<code><a href="generator_fasta_label_folder.html">generator_fasta_label_folder</a></code>, <code><a href="generator_fasta_label_header_csv.html">generator_fasta_label_header_csv</a></code>,
<code><a href="generator_rds.html">generator_rds</a></code>, <code><a href="generator_random.html">generator_random</a></code>, <code><a href="generator_dummy.html">generator_dummy</a></code> or
<code><a href="generator_fasta_lm.html">generator_fasta_lm</a></code> according to the <code>train_type</code> and <code>random_sampling</code>
arguments.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_generator</span><span class="op">(</span></span>
<span>  path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">train_type</span>,</span>
<span>  <span class="va">batch_size</span>,</span>
<span>  maxlen <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  step <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  shuffle_file_order <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  vocabulary <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"C"</span>, <span class="st">"G"</span>, <span class="st">"T"</span><span class="op">)</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  proportion_entries <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  shuffle_input <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  format <span class="op">=</span> <span class="st">"fasta"</span>,</span>
<span>  path_file_log <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  reverse_complement <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  n_gram <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  n_gram_stride <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  output_format <span class="op">=</span> <span class="st">"target_right"</span>,</span>
<span>  ambiguous_nuc <span class="op">=</span> <span class="st">"zero"</span>,</span>
<span>  proportion_per_seq <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  skip_amb_nuc <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  use_quality_score <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  padding <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  added_label_path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  target_from_csv <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  add_input_as_seq <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  max_samples <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  concat_seq <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  target_len <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  file_filter <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  use_coverage <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sample_by_file_size <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  add_noise <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  random_sampling <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  set_learning <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  file_limit <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  reverse_complement_encoding <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  read_data <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  split_seq <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  target_split <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  path_file_logVal <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  model <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  vocabulary_label <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  new_batch_size <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  val <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>path</dt>
<dd><p>Path to training data. If <code>train_type</code> is <code>label_folder</code>, should be a vector or list
where each entry corresponds to a class (list elements can be directories and/or individual files). If <code>train_type</code> is not <code>label_folder</code>,
can be a single directory or file or a list of directories and/or files.</p></dd>


<dt>train_type</dt>
<dd><p>Either <code>"lm"</code>, <code>"lm_rds"</code> for language model; <code>"label_header"</code>, <code>"label_folder"</code>, <code>"label_csv"</code>, <code>"label_rds"</code> for classification or <code>"dummy_gen"</code>.</p><ul><li><p>Language model is trained to predict character(s) in a sequence. <br></p></li>
<li><p><code>"label_header"</code>/<code>"label_folder"</code>/<code>"label_csv"</code> are trained to predict a corresponding class given a sequence as input.</p></li>
<li><p>If <code>"label_header"</code>, class will be read from fasta headers.</p></li>
<li><p>If <code>"label_folder"</code>, class will be read from folder, i.e. all files in one folder must belong to the same class.</p></li>
<li><p>If <code>"label_csv"</code>, targets are read from a csv file. This file should have one column named "file". The targets then correspond to entries in that row (except "file"
column). Example: if we are currently working with a file called "a.fasta" and corresponding label is "label_1", there should be a row in our csv file</p><table class="table table"><tr><td>file</td><td>label_1</td><td>label_2</td></tr><tr><td>"a.fasta"</td><td>1</td><td>0</td></tr></table></li>
<li><p>If <code>"label_rds"</code>, generator will iterate over set of .rds files containing each a list of input and target tensors. Not implemented for model
with multiple inputs.</p></li>
<li><p>If <code>"lm_rds"</code>, generator will iterate over set of .rds files and will split tensor according to <code>target_len</code> argument
(targets are last <code>target_len</code> nucleotides of each sequence).</p></li>
<li><p>If <code>"dummy_gen"</code>, generator creates random data once and repeatedly feeds these to model.</p></li>
</ul></dd>


<dt>batch_size</dt>
<dd><p>Number of samples used for one network update.</p></dd>


<dt>maxlen</dt>
<dd><p>Length of predictor sequence.</p></dd>


<dt>step</dt>
<dd><p>Frequency of sampling steps.</p></dd>


<dt>shuffle_file_order</dt>
<dd><p>Boolean, whether to go through files sequentially or shuffle beforehand.</p></dd>


<dt>vocabulary</dt>
<dd><p>Vector of allowed characters. Characters outside vocabulary get encoded as specified in <code>ambiguous_nuc</code>.</p></dd>


<dt>seed</dt>
<dd><p>Sets seed for reproducible results.</p></dd>


<dt>proportion_entries</dt>
<dd><p>Proportion of fasta entries to keep. For example, if fasta file has 50 entries and <code>proportion_entries = 0.1</code>,
will randomly select 5 entries.</p></dd>


<dt>shuffle_input</dt>
<dd><p>Whether to shuffle entries in file.</p></dd>


<dt>format</dt>
<dd><p>File format, <code>"fasta"</code>, <code>"fastq"</code> or <code>"rds"</code>.</p></dd>


<dt>path_file_log</dt>
<dd><p>Write name of files used for training to csv file if path is specified.</p></dd>


<dt>reverse_complement</dt>
<dd><p>Boolean, for every new file decide randomly to use original data or its reverse complement.</p></dd>


<dt>n_gram</dt>
<dd><p>Encode n nucleotides at once. Can be used for language model for target encoding.</p></dd>


<dt>n_gram_stride</dt>
<dd><p>Step size for n-gram encoding. For AACCGGTT with <code>n-gram = 4</code> and <code>n_gram_stride = 2</code>, generator encodes
<code>(AACC), (CCGG), (GGTT)</code>; for <code>n_gram_stride = 4</code> generator encodes <code>(AACC), (GGTT)</code>.</p></dd>


<dt>output_format</dt>
<dd><p>Determines shape of output tensor for language model.
Either <code>"target_right"</code>, <code>"target_middle_lstm"</code>, <code>"target_middle_cnn"</code> or <code>"wavenet"</code>.
Assume a sequence <code>"AACCGTA"</code>. Output correspond as follows</p><ul><li><p><code>"target_right": X = "AACCGT", Y = "A"</code></p></li>
<li><p><code>"target_middle_lstm": X = (X_1 = "AAC", X_2 = "ATG"), Y = "C"</code> (note reversed order of X_2)</p></li>
<li><p><code>"target_middle_cnn": X = "AACGTA", Y = "C"</code></p></li>
<li><p><code>"wavenet": X = "AACCGT", Y = "ACCGTA"</code></p></li>
</ul></dd>


<dt>ambiguous_nuc</dt>
<dd><p>How to handle nucleotides outside vocabulary, either <code>"zero"</code>, <code>"discard"</code>, <code>"empirical"</code> or <code>"equal"</code>.</p><ul><li><p>If <code>"zero"</code>, input gets encoded as zero vector.</p></li>
<li><p>If <code>"equal"</code>, input is repetition of <code>1/length(vocabulary)</code>.</p></li>
<li><p>If <code>"discard"</code>, samples containing nucleotides outside vocabulary get discarded.</p></li>
<li><p>If <code>"empirical"</code>, use nucleotide distribution of current file.</p></li>
</ul></dd>


<dt>proportion_per_seq</dt>
<dd><p>Numerical value between 0 and 1. Proportion of sequence to take samples from (use random subsequence).</p></dd>


<dt>skip_amb_nuc</dt>
<dd><p>Threshold of ambiguous nucleotides to accept in fasta entry. Complete entry will get discarded otherwise.</p></dd>


<dt>use_quality_score</dt>
<dd><p>Whether to use fastq quality scores. If <code>TRUE</code> input is not one-hot-encoding but corresponds to probabilities.
For example (0.97, 0.01, 0.01, 0.01) instead of (1, 0, 0, 0).</p></dd>


<dt>padding</dt>
<dd><p>Whether to pad sequences too short for one sample with zeros.</p></dd>


<dt>added_label_path</dt>
<dd><p>Path to file with additional input labels. Should be a csv file with one column named "file". Other columns should correspond to labels.</p></dd>


<dt>target_from_csv</dt>
<dd><p>Path to csv file with target mapping. One column should be called "file" and other entries in row are the targets.</p></dd>


<dt>add_input_as_seq</dt>
<dd><p>Boolean vector specifying for each entry in <code>added_label_path</code> if rows from csv should be encoded as a sequence or used directly.
If a row in your csv file is a sequence this should be <code>TRUE</code>. For example you may want to add another sequence, say ACCGT. Then this would correspond to 1,2,2,3,4 in
csv file (if vocabulary = c("A", "C", "G", "T")).  If <code>add_input_as_seq</code> is <code>TRUE</code>, 12234 gets one-hot encoded, so added input is a 3D tensor.  If <code>add_input_as_seq</code> is
<code>FALSE</code> this will feed network just raw data (a 2D tensor).</p></dd>


<dt>max_samples</dt>
<dd><p>Maximum number of samples to use from one file. If not <code>NULL</code> and file has more than <code>max_samples</code> samples, will randomly choose a
subset of <code>max_samples</code> samples.</p></dd>


<dt>concat_seq</dt>
<dd><p>Character string or <code>NULL</code>. If not <code>NULL</code> all entries from file get concatenated to one sequence with <code>concat_seq</code> string between them.
Example: If 1.entry AACC, 2. entry TTTG and <code>concat_seq = "ZZZ"</code> this becomes AACCZZZTTTG.</p></dd>


<dt>target_len</dt>
<dd><p>Number of nucleotides to predict at once for language model.</p></dd>


<dt>file_filter</dt>
<dd><p>Vector of file names to use from path_corpus.</p></dd>


<dt>use_coverage</dt>
<dd><p>Integer or <code>NULL</code>. If not <code>NULL</code>, use coverage as encoding rather than one-hot encoding and normalize.
Coverage information must be contained in fasta header: there must be a string <code>"cov_n"</code> in the header, where <code>n</code> is some integer.</p></dd>


<dt>sample_by_file_size</dt>
<dd><p>Sample new file weighted by file size (bigger files more likely).</p></dd>


<dt>add_noise</dt>
<dd><p><code>NULL</code> or list of arguments. If not <code>NULL</code>, list must contain the following arguments: <code>noise_type</code> can be <code>"normal"</code> or <code>"uniform"</code>;
optional arguments <code>sd</code> or <code>mean</code> if noise_type is <code>"normal"</code> (default is <code>sd=1</code> and <code>mean=0</code>) or <code>min, max</code> if <code>noise_type</code> is <code>"uniform"</code>
(default is <code>min=0, max=1</code>).</p></dd>


<dt>random_sampling</dt>
<dd><p>Whether samples should be taken from random positions when using <code>max_samples</code> argument. If <code>FALSE</code> random
samples are taken from a consecutive subsequence.</p><ul><li><p><code>samples_per_target</code>: how many samples to use for one target.</p></li>
<li><p><code>maxlen</code>: length of one sample.</p></li>
<li><p><code>reshape_mode</code>: <code>"time_dist", "multi_input"</code> or <code>"concat"</code>.</p><ul><li><p>If <code>reshape_mode</code> is <code>"multi_input"</code>, generator will produce <code>samples_per_target</code> separate inputs, each of length <code>maxlen</code> (model should have
<code>samples_per_target</code> input layers).</p></li>
<li><p>If reshape_mode is <code>"time_dist"</code>, generator will produce a 4D input array. The dimensions correspond to
<code>(batch_size, samples_per_target, maxlen, length(vocabulary))</code>.</p></li>
<li><p>If <code>reshape_mode</code> is <code>"concat"</code>, generator will concatenate <code>samples_per_target</code> sequences
of length <code>maxlen</code> to one long sequence.</p></li>
</ul></li>
<li><p>If <code>reshape_mode</code> is <code>"concat"</code>, there is an additional <code>buffer_len</code>
argument. If <code>buffer_len</code> is an integer, the subsequences are interspaced with <code>buffer_len</code> rows. The input length is
(<code>maxlen</code> \(*\) <code>samples_per_target</code>) + <code>buffer_len</code> \(*\) (<code>samples_per_target</code> - 1).</p></li>
</ul></dd>


<dt>set_learning</dt>
<dd><p>When you want to assign one label to set of samples. Only implemented for <code>train_type = "label_folder"</code>.
Input is a list with the following parameters</p></dd>


<dt>file_limit</dt>
<dd><p>Integer or <code>NULL</code>. If integer, use only specified number of randomly sampled files for training. Ignored if greater than number of files in <code>path</code>.</p></dd>


<dt>reverse_complement_encoding</dt>
<dd><p>Whether to use both original sequence and reverse complement as two input sequences.</p></dd>


<dt>read_data</dt>
<dd><p>If <code>TRUE</code> the first element of output is a list of length 2, each containing one part of paired read. Maxlen should be 2*length of one read.</p></dd>


<dt>split_seq</dt>
<dd><p>Split input sequence into two sequences while removing nucleotide in middle. If input is <code>x_1,..., x_(n+1)</code>, input gets split into
<code>input_1 = x_1,..., x_m</code> and <code>input_2 = x_(n+1),..., x_(m+2)</code> where <code>m = ceiling((n+1)/2)</code> and <code>n = maxlen</code>. Note that <code>x_(m+1)</code> is not used. Can be used for transfer learning,
when switching from language model trained with target in middle to label classification.</p></dd>


<dt>target_split</dt>
<dd><p>If target gets read from csv file, list of names to divide target tensor into list of tensors.
Example: if csv file has header names <code>"file", "label_1", "label_2", "label_3"</code> and <code>target_split = list(c("label_1", "label_2"), "label_3")</code>,
this will divide target matrix to list of length 2, where the first element contains columns named <code>"label_1"</code> and <code>"label_2"</code> and the
second entry contains the column named <code>"label_3"</code>.</p></dd>


<dt>path_file_logVal</dt>
<dd><p>Path to csv file logging used validation files.</p></dd>


<dt>model</dt>
<dd><p>A keras model.</p></dd>


<dt>vocabulary_label</dt>
<dd><p>Character vector of possible targets. Targets outside <code>vocabulary_label</code> will get discarded if
<code>train_type = "label_header"</code>.</p></dd>


<dt>new_batch_size</dt>
<dd><p>Batch size after reshaping data for set learning.</p></dd>


<dt>val</dt>
<dd><p>Logical, call initialized generator "genY" or "genValY" where Y is an integer between 1 and length of directories.</p></dd>

</dl></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># create dummy fasta files</span></span></span>
<span class="r-in"><span><span class="va">fasta_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">fasta_path</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="create_dummy_data.html">create_dummy_data</a></span><span class="op">(</span>file_path <span class="op">=</span> <span class="va">fasta_path</span>,</span></span>
<span class="r-in"><span>                  num_files <span class="op">=</span> <span class="fl">3</span>,</span></span>
<span class="r-in"><span>                  seq_length <span class="op">=</span> <span class="fl">10</span>,</span></span>
<span class="r-in"><span>                  num_seq <span class="op">=</span> <span class="fl">5</span>,</span></span>
<span class="r-in"><span>                  vocabulary <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a"</span>, <span class="st">"c"</span>, <span class="st">"g"</span>, <span class="st">"t"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">gen</span> <span class="op">&lt;-</span> <span class="fu">get_generator</span><span class="op">(</span>path <span class="op">=</span> <span class="va">fasta_path</span>,</span></span>
<span class="r-in"><span>                     maxlen <span class="op">=</span> <span class="fl">5</span>, train_type <span class="op">=</span> <span class="st">"lm"</span>,</span></span>
<span class="r-in"><span>                     output_format <span class="op">=</span> <span class="st">"target_right"</span>,</span></span>
<span class="r-in"><span>                     step <span class="op">=</span> <span class="fl">3</span>, batch_size <span class="op">=</span> <span class="fl">7</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">z</span> <span class="op">&lt;-</span> <span class="fu">gen</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">z</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">z</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 7 5 4</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 7 4</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Philipp Münch, René Mreches, Martin Binder, Hüseyin Anil Gündüz, Alice McHardy.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer></div>

  


  

  </body></html>

