<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference • deepG</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><meta property="og:image" content="https://genomenet.github.io/deepG/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">deepG</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.1-9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">
    <span class="fa fa-file-alt"></span>
     
    Reference
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Notebooks
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="https://colab.research.google.com/drive/175jIdXcDcgPUvaBo2rH2Lupbpjnp5O7G?usp=sharing" class="external-link">deepG tutorial</a>
    </li>
    <li>
      <a href="https://colab.research.google.com/drive/1Eolc0koMNM1zkuO4XyVM58ImeF1BpRiH?usp=sharing" class="external-link">Read-length level: Human contamination</a>
    </li>
    <li>
      <a href="https://colab.research.google.com/drive/1yiXSwFafXpMLHaov9iBTQLIDZ6bK1zYX?usp=sharing" class="external-link">Locus level: CRISPR detection</a>
    </li>
    <li>
      <a href="https://colab.research.google.com/drive/1G7bOFEX87cZNrM2tdRtTdkrZn5fM__g0?usp=sharing" class="external-link">Gene level: 16S rRNA detection</a>
    </li>
    <li>
      <a href="https://colab.research.google.com/drive/1BCggL-tfQF136YeJ8cKKi-zoBEDMgkNh?usp=sharing" class="external-link">Genome level: Bacterial morphology (Sporulation)</a>
    </li>
    <li>
      <a href="https://colab.research.google.com/drive/10xpRzGd3JeBAbqQYSCxzQUMctt01sx9D?usp=sharing" class="external-link">Full metagenome level: Colorectal cancer prediction</a>
    </li>
    <li>
      <a href="https://colab.research.google.com/drive/1kyYK7IU7GSfdpDzO_a8U3_qD4i3zTu6w?usp=sharing" class="external-link">BERT with deepG</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/getting_started.html">Getting Started</a>
    </li>
    <li>
      <a href="../articles/training_types.html">Training types</a>
    </li>
    <li>
      <a href="../articles/data_generator.html">Data generator</a>
    </li>
    <li>
      <a href="../articles/using_tb.html">Using tensorboard</a>
    </li>
    <li>
      <a href="../articles/integrated_gradient.html">Integrated Gradient</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/GenomeNet/deepG/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="training">Training <a href="#training" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="train_model.html">train_model()</a></code> </p>
        </td>
        <td><p>Train neural network on genomic data</p></td>
      </tr><tr><td>
          <p><code><a href="train_model_cpc.html">train_model_cpc()</a></code> </p>
        </td>
        <td><p>Train CPC inspired model</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="model-architecture">Model Architecture <a href="#model-architecture" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="create_model_genomenet.html">create_model_genomenet()</a></code> </p>
        </td>
        <td><p>Create GenomeNet Model with Given Architecture Parameters</p></td>
      </tr><tr><td>
          <p><code><a href="create_model_lstm_cnn.html">create_model_lstm_cnn()</a></code> </p>
        </td>
        <td><p>Create LSTM/CNN network</p></td>
      </tr><tr><td>
          <p><code><a href="create_model_lstm_cnn_multi_input.html">create_model_lstm_cnn_multi_input()</a></code> </p>
        </td>
        <td><p>Create LSTM/CNN network that can process multiple samples for one target</p></td>
      </tr><tr><td>
          <p><code><a href="create_model_lstm_cnn_target_middle.html">create_model_lstm_cnn_target_middle()</a></code> </p>
        </td>
        <td><p>Create LSTM/CNN network to predict middle part of a sequence</p></td>
      </tr><tr><td>
          <p><code><a href="create_model_lstm_cnn_time_dist.html">create_model_lstm_cnn_time_dist()</a></code> </p>
        </td>
        <td><p>Create LSTM/CNN network for combining multiple sequences</p></td>
      </tr><tr><td>
          <p><code><a href="create_model_transformer.html">create_model_transformer()</a></code> </p>
        </td>
        <td><p>Create transformer model</p></td>
      </tr><tr><td>
          <p><code><a href="create_model_twin_network.html">create_model_twin_network()</a></code> </p>
        </td>
        <td><p>Create twin network with contrastive loss</p></td>
      </tr><tr><td>
          <p><code><a href="create_model_wavenet.html">create_model_wavenet()</a></code> </p>
        </td>
        <td><p>Create wavenet model</p></td>
      </tr><tr><td>
          <p><code><a href="remove_add_layers.html">remove_add_layers()</a></code> </p>
        </td>
        <td><p>Remove layers from model and add dense layers</p></td>
      </tr><tr><td>
          <p><code><a href="merge_models.html">merge_models()</a></code> </p>
        </td>
        <td><p>Merge two models</p></td>
      </tr><tr><td>
          <p><code><a href="get_output_activations.html">get_output_activations()</a></code> </p>
        </td>
        <td><p>Get activation functions of output layers</p></td>
      </tr><tr><td>
          <p><code><a href="reshape_input.html">reshape_input()</a></code> </p>
        </td>
        <td><p>Replace input layer</p></td>
      </tr><tr><td>
          <p><code><a href="load_cp.html">load_cp()</a></code> </p>
        </td>
        <td><p>Load checkpoint</p></td>
      </tr><tr><td>
          <p><code><a href="compile_model.html">compile_model()</a></code> </p>
        </td>
        <td><p>Compile model</p></td>
      </tr><tr><td>
          <p><code><a href="layer_aggregate_time_dist_wrapper.html">layer_aggregate_time_dist_wrapper()</a></code> </p>
        </td>
        <td><p>Add layer</p></td>
      </tr><tr><td>
          <p><code><a href="layer_pos_embedding_wrapper.html">layer_pos_embedding_wrapper()</a></code> </p>
        </td>
        <td><p>Layer for positional embedding</p></td>
      </tr><tr><td>
          <p><code><a href="layer_pos_sinusoid_wrapper.html">layer_pos_sinusoid_wrapper()</a></code> </p>
        </td>
        <td><p>Layer for positional encoding</p></td>
      </tr><tr><td>
          <p><code><a href="layer_transformer_block_wrapper.html">layer_transformer_block_wrapper()</a></code> </p>
        </td>
        <td><p>Transformer block</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="callbacks">Callbacks <a href="#callbacks" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="conf_matrix_cb.html">conf_matrix_cb()</a></code> </p>
        </td>
        <td><p>Confusion matrix callback.</p></td>
      </tr><tr><td>
          <p><code><a href="early_stopping_time_cb.html">early_stopping_time_cb()</a></code> </p>
        </td>
        <td><p>Stop training callback</p></td>
      </tr><tr><td>
          <p><code><a href="function_args_cb.html">function_args_cb()</a></code> </p>
        </td>
        <td><p>Function arguments callback</p></td>
      </tr><tr><td>
          <p><code><a href="validation_after_training_cb.html">validation_after_training_cb()</a></code> </p>
        </td>
        <td><p>Validation after training callback</p></td>
      </tr><tr><td>
          <p><code><a href="reset_states_cb.html">reset_states_cb()</a></code> </p>
        </td>
        <td><p>Reset states callback</p></td>
      </tr><tr><td>
          <p><code><a href="remove_checkpoints.html">remove_checkpoints()</a></code> </p>
        </td>
        <td><p>Remove checkpoints</p></td>
      </tr><tr><td>
          <p><code><a href="model_card_cb.html">model_card_cb()</a></code> </p>
        </td>
        <td><p>Create model card</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="metrics-and-losses">Metrics and Losses <a href="#metrics-and-losses" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="noisy_loss_wrapper.html">noisy_loss_wrapper()</a></code> </p>
        </td>
        <td><p>Loss function for label noise</p></td>
      </tr><tr><td>
          <p><code><a href="balanced_acc_wrapper.html">balanced_acc_wrapper()</a></code> </p>
        </td>
        <td><p>Balanced accuracy metric</p></td>
      </tr><tr><td>
          <p><code><a href="f1_wrapper.html">f1_wrapper()</a></code> </p>
        </td>
        <td><p>F1 metric</p></td>
      </tr><tr><td>
          <p><code><a href="auc_wrapper.html">auc_wrapper()</a></code> </p>
        </td>
        <td><p>Mean AUC score</p></td>
      </tr><tr><td>
          <p><code><a href="exp_decay.html">exp_decay()</a></code> </p>
        </td>
        <td><p>Exponential Decay</p></td>
      </tr><tr><td>
          <p><code><a href="stepdecay.html">stepdecay()</a></code> </p>
        </td>
        <td><p>Step Decay</p></td>
      </tr><tr><td>
          <p><code><a href="sgdr.html">sgdr()</a></code> </p>
        </td>
        <td><p>Stochastic Gradient Descent with Warm Restarts</p></td>
      </tr><tr><td>
          <p><code><a href="focal_loss_multiclass.html">focal_loss_multiclass()</a></code> </p>
        </td>
        <td><p>Focal loss for two or more labels</p></td>
      </tr><tr><td>
          <p><code><a href="loss_cl.html">loss_cl()</a></code> </p>
        </td>
        <td><p>Contrastive loss</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="generators">Generators <a href="#generators" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="generator_dummy.html">generator_dummy()</a></code> </p>
        </td>
        <td><p>Random data generator</p></td>
      </tr><tr><td>
          <p><code><a href="generator_fasta_label_folder.html">generator_fasta_label_folder()</a></code> </p>
        </td>
        <td><p>Data generator for fasta/fasta files</p></td>
      </tr><tr><td>
          <p><code><a href="generator_fasta_label_folder_wrapper.html">generator_fasta_label_folder_wrapper()</a></code> </p>
        </td>
        <td><p>Generator wrapper</p></td>
      </tr><tr><td>
          <p><code><a href="generator_fasta_label_header_csv.html">generator_fasta_label_header_csv()</a></code> </p>
        </td>
        <td><p>Data generator for fasta/fastq files and label targets</p></td>
      </tr><tr><td>
          <p><code><a href="generator_fasta_lm.html">generator_fasta_lm()</a></code> </p>
        </td>
        <td><p>Language model generator for fasta/fastq files</p></td>
      </tr><tr><td>
          <p><code><a href="generator_initialize.html">generator_initialize()</a></code> </p>
        </td>
        <td><p>Initializes generators defined by <code>generator_fasta_label_folder</code> function</p></td>
      </tr><tr><td>
          <p><code><a href="generator_random.html">generator_random()</a></code> </p>
        </td>
        <td><p>Randomly select samples from fasta files</p></td>
      </tr><tr><td>
          <p><code><a href="generator_rds.html">generator_rds()</a></code> </p>
        </td>
        <td><p>Rds data generator</p></td>
      </tr><tr><td>
          <p><code><a href="get_generator.html">get_generator()</a></code> </p>
        </td>
        <td><p>Wrapper for generator functions</p></td>
      </tr><tr><td>
          <p><code><a href="dataset_from_gen.html">dataset_from_gen()</a></code> </p>
        </td>
        <td><p>Collect samples from generator and store in rds or pickle file.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="prediction">Prediction <a href="#prediction" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="predict_model.html">predict_model()</a></code> </p>
        </td>
        <td><p>Make prediction for nucleotide sequence or entries in fasta/fastq file</p></td>
      </tr><tr><td>
          <p><code><a href="summarize_states.html">summarize_states()</a></code> </p>
        </td>
        <td><p>Create summary of predictions</p></td>
      </tr><tr><td>
          <p><code><a href="predict_with_n_gram.html">predict_with_n_gram()</a></code> </p>
        </td>
        <td><p>Predict the next nucleotide using n-gram</p></td>
      </tr><tr><td>
          <p><code><a href="load_prediction.html">load_prediction()</a></code> </p>
        </td>
        <td><p>Read states from h5 file</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="evaluation">Evaluation <a href="#evaluation" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="evaluate_linear.html">evaluate_linear()</a></code> </p>
        </td>
        <td><p>Evaluate matrices of true targets and predictions from layer with linear activation.</p></td>
      </tr><tr><td>
          <p><code><a href="evaluate_model.html">evaluate_model()</a></code> </p>
        </td>
        <td><p>Evaluates a trained model on fasta, fastq or rds files</p></td>
      </tr><tr><td>
          <p><code><a href="evaluate_sigmoid.html">evaluate_sigmoid()</a></code> </p>
        </td>
        <td><p>Evaluate matrices of true targets and predictions from layer with sigmoid activation.</p></td>
      </tr><tr><td>
          <p><code><a href="evaluate_softmax.html">evaluate_softmax()</a></code> </p>
        </td>
        <td><p>Evaluate matrices of true targets and predictions from layer with softmax activation.</p></td>
      </tr><tr><td>
          <p><code><a href="plot_roc.html">plot_roc()</a></code> </p>
        </td>
        <td><p>Plot ROC</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="visualization">Visualization <a href="#visualization" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="integrated_gradients.html">integrated_gradients()</a></code> </p>
        </td>
        <td><p>Compute integrated gradients</p></td>
      </tr><tr><td>
          <p><code><a href="heatmaps_integrated_grad.html">heatmaps_integrated_grad()</a></code> </p>
        </td>
        <td><p>Heatmap of integrated gradient scores</p></td>
      </tr><tr><td>
          <p><code><a href="plot_cm.html">plot_cm()</a></code> </p>
        </td>
        <td><p>Plot confusion matrix</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="preprocessing">Preprocessing <a href="#preprocessing" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="seq_encoding_label.html">seq_encoding_label()</a></code> </p>
        </td>
        <td><p>Encodes integer sequence for label classification.</p></td>
      </tr><tr><td>
          <p><code><a href="seq_encoding_lm.html">seq_encoding_lm()</a></code> </p>
        </td>
        <td><p>Encodes integer sequence for language model</p></td>
      </tr><tr><td>
          <p><code><a href="get_class_weight.html">get_class_weight()</a></code> </p>
        </td>
        <td><p>Estimate frequency of different classes</p></td>
      </tr><tr><td>
          <p><code><a href="get_start_ind.html">get_start_ind()</a></code> </p>
        </td>
        <td><p>Computes start position of samples</p></td>
      </tr><tr><td>
          <p><code><a href="int_to_n_gram.html">int_to_n_gram()</a></code> </p>
        </td>
        <td><p>Encode sequence of integers to sequence of n-gram</p></td>
      </tr><tr><td>
          <p><code><a href="n_gram_dist.html">n_gram_dist()</a></code> </p>
        </td>
        <td><p>Get distribution of n-grams</p></td>
      </tr><tr><td>
          <p><code><a href="n_gram_of_matrix.html">n_gram_of_matrix()</a></code> </p>
        </td>
        <td><p>One-hot encoding matrix to n-gram encoding matrix</p></td>
      </tr><tr><td>
          <p><code><a href="reshape_tensor.html">reshape_tensor()</a></code> </p>
        </td>
        <td><p>Reshape tensors for set learning</p></td>
      </tr><tr><td>
          <p><code><a href="split_fasta.html">split_fasta()</a></code> </p>
        </td>
        <td><p>Split fasta file into smaller files.</p></td>
      </tr><tr><td>
          <p><code><a href="one_hot_to_seq.html">one_hot_to_seq()</a></code> </p>
        </td>
        <td><p>Char sequence corresponding to one-hot matrix.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="data">Data <a href="#data" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="crispr_full.html">crispr_full</a></code> </p>
        </td>
        <td><p>CRISPR data subset</p></td>
      </tr><tr><td>
          <p><code><a href="crispr_sample.html">crispr_sample</a></code> </p>
        </td>
        <td><p>CRISPR data</p></td>
      </tr><tr><td>
          <p><code><a href="ecoli.html">ecoli</a></code> </p>
        </td>
        <td><p>Ecoli genome</p></td>
      </tr><tr><td>
          <p><code><a href="ecoli_small.html">ecoli_small</a></code> </p>
        </td>
        <td><p>Ecoli subset</p></td>
      </tr><tr><td>
          <p><code><a href="parenthesis.html">parenthesis</a></code> </p>
        </td>
        <td><p>Parenthesis data</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="dummy-data">Dummy Data <a href="#dummy-data" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="create_dummy_data.html">create_dummy_data()</a></code> </p>
        </td>
        <td><p>Write random sequences to fasta file</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="session-management">Session Management <a href="#session-management" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="deepG-package.html">deepG</a></code> <code><a href="deepG-package.html">deepG-package</a></code> </p>
        </td>
        <td><p>deepG for GenomeNet</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Philipp Münch, René Mreches, Martin Binder, Hüseyin Anil Gündüz, Xiao-Yin To, Alice McHardy.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

