<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Make prediction for nucleotide sequence or entries in fasta/fastq file — predict_model • deepG</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Make prediction for nucleotide sequence or entries in fasta/fastq file — predict_model"><meta property="og:description" content='Removes layers (optional) from pretrained model and calculates states of fasta/fastq file or nucleotide sequence.
Writes states to h5 or csv file (access content of h5 output with load_predictions function).
There are several options on how to process an input file:
If "one_seq", computes prediction for sequence argument or fasta/fastq file.
Combines fasta entries in file to one sequence. This means predictor sequences can contain elements from more than one fasta entry.
If "by_entry", will output a separate file for each fasta/fastq entry.
Names of output files are: output_dir + "Nr" + i + filename + output_type, where i is the number of the fasta entry.
If "by_entry_one_file", will store prediction for all fasta entries in one h5 file.
If "one_pred_per_entry", will make one prediction for each entry by either picking random sample for long sequences
or pad sequence for short sequences.

'><meta property="og:image" content="https://genomenet.github.io/deepG/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">deepG</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.1-9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">
    <span class="fa fa-file-alt"></span>
     
    Reference
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Google Colab Notebooks
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="https://colab.research.google.com/drive/175jIdXcDcgPUvaBo2rH2Lupbpjnp5O7G?usp=sharing" class="external-link">deepG tutorial</a>
    </li>
    <li>
      <a href="https://colab.research.google.com/drive/1Eolc0koMNM1zkuO4XyVM58ImeF1BpRiH?usp=sharing" class="external-link">Human contamination</a>
    </li>
    <li>
      <a href="https://colab.research.google.com/drive/1MmQB2pFfEcp4GbgbT7wu6ScNt1YsChmn?usp=sharing" class="external-link">CRISPR detection</a>
    </li>
    <li>
      <a href="https://colab.research.google.com/drive/1G7bOFEX87cZNrM2tdRtTdkrZn5fM__g0?usp=sharing" class="external-link">16S rRNA detection</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/getting_started.html">Getting Started</a>
    </li>
    <li>
      <a href="../articles/training_types.html">Training types</a>
    </li>
    <li>
      <a href="../articles/data_generator.html">Data generator</a>
    </li>
    <li>
      <a href="../articles/using_tb.html">Using tensorboard</a>
    </li>
    <li>
      <a href="../articles/integrated_gradient.html">Integrated Gradient</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/GenomeNet/deepG/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Make prediction for nucleotide sequence or entries in fasta/fastq file</h1>
    <small class="dont-index">Source: <a href="https://github.com/GenomeNet/deepG/blob/HEAD/R/predict.R" class="external-link"><code>R/predict.R</code></a></small>
    <div class="hidden name"><code>predict_model.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Removes layers (optional) from pretrained model and calculates states of fasta/fastq file or nucleotide sequence.
Writes states to h5 or csv file (access content of h5 output with <code>load_predictions</code> function).
There are several options on how to process an input file:</p><ul><li><p>If <code>"one_seq"</code>, computes prediction for sequence argument or fasta/fastq file.
Combines fasta entries in file to one sequence. This means predictor sequences can contain elements from more than one fasta entry.</p></li>
<li><p>If <code>"by_entry"</code>, will output a separate file for each fasta/fastq entry.
Names of output files are: <code>output_dir</code> + "Nr" + i + <code>filename</code> + <code>output_type</code>, where i is the number of the fasta entry.</p></li>
<li><p>If <code>"by_entry_one_file"</code>, will store prediction for all fasta entries in one h5 file.</p></li>
<li><p>If <code>"one_pred_per_entry"</code>, will make one prediction for each entry by either picking random sample for long sequences
or pad sequence for short sequences.</p></li>
</ul></div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">predict_model</span><span class="op">(</span></span>
<span>  output_format <span class="op">=</span> <span class="st">"one_seq"</span>,</span>
<span>  model <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  layer_name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sequence <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  path_input <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  round_digits <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  filename <span class="op">=</span> <span class="st">"states.h5"</span>,</span>
<span>  step <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  vocabulary <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a"</span>, <span class="st">"c"</span>, <span class="st">"g"</span>, <span class="st">"t"</span><span class="op">)</span>,</span>
<span>  batch_size <span class="op">=</span> <span class="fl">256</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  return_states <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  padding_maxlen <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  output_type <span class="op">=</span> <span class="st">"h5"</span>,</span>
<span>  path_model <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"lm"</span>,</span>
<span>  lm_format <span class="op">=</span> <span class="st">"target_right"</span>,</span>
<span>  output_dir <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  format <span class="op">=</span> <span class="st">"fasta"</span>,</span>
<span>  include_seq <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  reverse_complement_encoding <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>output_format</dt>
<dd><p>Either <code>"one_seq"</code>, <code>"by_entry"</code>, <code>"by_entry_one_file"</code>, <code>"one_pred_per_entry"</code>.</p></dd>


<dt>model</dt>
<dd><p>A keras model. If model and path_model are not NULL, model will be used for inference.</p></dd>


<dt>layer_name</dt>
<dd><p>Name of layer to get output from. If <code>NULL</code>, will use the last layer.</p></dd>


<dt>sequence</dt>
<dd><p>Character string, ignores path_input if argument given.</p></dd>


<dt>path_input</dt>
<dd><p>Path to fasta file.</p></dd>


<dt>round_digits</dt>
<dd><p>Number of decimal places.</p></dd>


<dt>filename</dt>
<dd><p>Filename to store states in. No file output if argument is <code>NULL</code>.</p></dd>


<dt>step</dt>
<dd><p>Frequency of sampling steps.</p></dd>


<dt>vocabulary</dt>
<dd><p>Vector of allowed characters, character outside vocabulary get encoded as 0-vector.</p></dd>


<dt>batch_size</dt>
<dd><p>Number of samples to evaluate at once. Does not change output, only relevant for speed and memory.</p></dd>


<dt>verbose</dt>
<dd><p>Boolean.</p></dd>


<dt>return_states</dt>
<dd><p>Return predictions as data frame. Only supported for output_format <code>"one_seq"</code>.</p></dd>


<dt>padding_maxlen</dt>
<dd><p>Boolean, whether to pad input with maxlen 0-vectors. (First prediction is for tensor of zeros)</p></dd>


<dt>output_type</dt>
<dd><p><code>"h5"</code> or <code>"csv"</code>. If <code>output_format`` is </code>"by_entries_one_file", "one_pred_per_entry"<code>can only be</code>"h5"`.</p></dd>


<dt>path_model</dt>
<dd><p>Path to a pretrained model.</p></dd>


<dt>mode</dt>
<dd><p>Either <code>"lm"</code> for language model or <code>"label"</code> for label classification.</p></dd>


<dt>output_dir</dt>
<dd><p>Path to folder, where to write output.</p></dd>


<dt>format</dt>
<dd><p>Either <code>"fasta"</code> or <code>"fastq"</code>.</p></dd>


<dt>include_seq</dt>
<dd><p>Whether to include input sequence in h5 file.</p></dd>

</dl></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># make prediction for single sequence and write to h5 file</span></span></span>
<span class="r-in"><span><span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="create_model_lstm_cnn.html">create_model_lstm_cnn</a></span><span class="op">(</span>maxlen <span class="op">=</span> <span class="fl">20</span>, layer_lstm <span class="op">=</span> <span class="fl">8</span>, layer_dense <span class="op">=</span> <span class="fl">2</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">vocabulary</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a"</span>, <span class="st">"c"</span>, <span class="st">"g"</span>, <span class="st">"t"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sequence</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">vocabulary</span>, <span class="fl">200</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">output_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>fileext <span class="op">=</span> <span class="st">".h5"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">predict_model</span><span class="op">(</span>output_format <span class="op">=</span> <span class="st">"one_seq"</span>, model <span class="op">=</span> <span class="va">model</span>, step <span class="op">=</span> <span class="fl">10</span>,</span></span>
<span class="r-in"><span>             sequence <span class="op">=</span> <span class="va">sequence</span>, filename <span class="op">=</span> <span class="va">output_file</span>, mode <span class="op">=</span> <span class="st">"label"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> layer_name not specified. Using layer dense_25</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Computing output for model at layer dense_25 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Model: "model_21"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ________________________________________________________________________________</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Layer (type)                       Output Shape                    Param #     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  input_17 (InputLayer)              [(None, 20, 4)]                 0           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  lstm_19 (LSTM)                     (None, 8)                       416         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  dense_25 (Dense)                   (None, 2)                       18          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total params: 434</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Trainable params: 434</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Non-trainable params: 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ________________________________________________________________________________</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NULL</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># make prediction for fasta file with multiple entries, write output to separate h5 files</span></span></span>
<span class="r-in"><span><span class="va">fasta_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>fileext <span class="op">=</span> <span class="st">".fasta"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="create_dummy_data.html">create_dummy_data</a></span><span class="op">(</span>file_path <span class="op">=</span> <span class="va">fasta_path</span>, num_files <span class="op">=</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>                 num_seq <span class="op">=</span> <span class="fl">5</span>, seq_length <span class="op">=</span> <span class="fl">100</span>,</span></span>
<span class="r-in"><span>                 write_to_file_path <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="create_model_lstm_cnn.html">create_model_lstm_cnn</a></span><span class="op">(</span>maxlen <span class="op">=</span> <span class="fl">20</span>, layer_lstm <span class="op">=</span> <span class="fl">8</span>, layer_dense <span class="op">=</span> <span class="fl">2</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">output_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">output_dir</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">predict_model</span><span class="op">(</span>output_format <span class="op">=</span> <span class="st">"by_entry"</span>, model <span class="op">=</span> <span class="va">model</span>, step <span class="op">=</span> <span class="fl">10</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>               output_dir <span class="op">=</span> <span class="va">output_dir</span>, mode <span class="op">=</span> <span class="st">"label"</span>, path_input <span class="op">=</span> <span class="va">fasta_path</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">output_dir</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "states_nr_1.h5" "states_nr_2.h5" "states_nr_3.h5" "states_nr_4.h5"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [5] "states_nr_5.h5"</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Philipp Münch, René Mreches, Martin Binder, Hüseyin Anil Gündüz, Alice McHardy.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer></div>

  


  

  </body></html>

